---
title: "Differential analysis of bulk CUT&Tag data"
author: "Samantha Lee"
date: "`r Sys.Date()`"
output:
  html_document:
    code_folding: hide
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE,
                      message = FALSE,
                      warning = FALSE)
```

## Load data 

Load required libraries
```{r req_libs}
library(here) # dir calling
library(tidyverse) # data wrangling
library(sva) # surrogate variables
library(DESeq2) # diff analysis
library(edgeR) # diff analysis 
library(ChIPseeker) # peak anno
library(clusterProfiler) # go analysis
library(TxDb.Hsapiens.UCSC.hg38.knownGene) # anno library
library(org.Hs.eg.db) # human annotation
library(Gviz) # ucsc track vis
```

Source required custom function
```{r src}
# qqplot and lambda
source(here("r",
            "qqplot.r"))
```

Read in participant metadata.
```{r data_in_meta}
meta <- read.csv(here("resources",
                      "metadata",
                      "metadata_participants.csv")) %>%
  mutate(group = ifelse(grepl("PDC", pid), "control", "pd")) %>%
  mutate(sample = str_extract(sample, "IGF\\d+"))
```

Read in FRiP scores. 
```{r data_in_frip_scores}
frip_in <- function(f) {
  read_table(f,
             skip = 0, 
             col_names = T)
}

frip_scores <- map_dfr(list.files(here("data_out",
                                       "070_peaks", 
                                       "072_frip_scores"),
                                  pattern = "^.*\\.txt$",
                                  full.names = TRUE), 
                       frip_in) 
```

Combine metadata with FRiP score.
```{r combine_meta_frip}
meta <- merge(meta,
              frip_scores %>%
                dplyr::select(file, percent) %>%
                dplyr::rename("frip" = percent,
                              "sample" = file),
              by = "sample")
```

Read in celltype-specific feature counts.
```{r data_in_feature_counts}
micro_tsv_files <- list.files(here("data_out",
                                   "070_peaks",
                                   "078_peaks_feature_counts_celltype"), 
                              pattern = "*microglia*", 
                              full.names = TRUE)

oligo_tsv_files <- list.files(here("data_out",
                                   "070_peaks",
                                   "078_peaks_feature_counts_celltype"), 
                              pattern = "*oligodendrocyte*", 
                              full.names = TRUE)

neuro_tsv_files <- list.files(here("data_out",
                                   "070_peaks",
                                   "078_peaks_feature_counts_celltype"), 
                              pattern = "*neuron*", 
                              full.names = TRUE)

micro_feat <- purrr::imap_dfc(
  micro_tsv_files,
  ~ read_tsv(.x, show_col_types = FALSE, col_names = FALSE) %>% 
    { if (.y == 1) dplyr::select(., 1:3, dplyr::last_col())
      else          dplyr::select(., dplyr::last_col()) }
)


oligo_feat <- purrr::imap_dfc(
  oligo_tsv_files,
  ~ read_tsv(.x, show_col_types = FALSE, col_names = FALSE) %>% 
    { if (.y == 1) dplyr::select(., 1:3, dplyr::last_col())
      else          dplyr::select(., dplyr::last_col()) }
)

neuro_feat <- purrr::imap_dfc(
  neuro_tsv_files,
  ~ read_tsv(.x, show_col_types = FALSE, col_names = FALSE) %>% 
    { if (.y == 1) dplyr::select(., 1:3, dplyr::last_col())
      else          dplyr::select(., dplyr::last_col()) }
)

colnames(micro_feat) <- c("chr", "start", "end",
                          sub("_microglia_multicov\\.tsv$", 
                              "", 
                              basename(micro_tsv_files)))

colnames(oligo_feat) <- c("chr", "start", "end",
                          sub("_oligodendrocytes_multicov\\.tsv$", 
                              "", 
                              basename(oligo_tsv_files)))

colnames(neuro_feat) <- c("chr", "start", "end",
                          sub("_neuron_multicov\\.tsv$", 
                              "", 
                              basename(neuro_tsv_files)))

# remove low frip samples
micro_feat_clean <- micro_feat %>%
  dplyr::select(!"IGF136817")

oligo_feat_clean <- oligo_feat %>%
  dplyr::select(!"IGF136816")

neuro_feat_clean <- neuro_feat %>%
  dplyr::select(!"IGF136815")

# remove uneeded files 
rm(micro_tsv_files,
   oligo_tsv_files,
   neuro_tsv_files,
   micro_feat,
   oligo_feat,
   neuro_feat,
   frip_scores)
```

Subset metadata for each cell type
```{r meta_celltype}
micro_meta <- meta %>%
  mutate(sample = str_extract(sample, "IGF\\d+")) %>%
  subset(sample %in% colnames(micro_feat_clean))

oligo_meta <- meta %>%
  mutate(sample = str_extract(sample, "IGF\\d+")) %>%
  subset(sample %in% colnames(oligo_feat_clean))

neuro_meta <- meta %>%
  mutate(sample = str_extract(sample, "IGF\\d+")) %>%
  subset(sample %in% colnames(neuro_feat_clean))
```

## Annotate peaks for each cell type

```{r micro_peak_anno}
# convert peaks to granges obj
micro_feat_gr  <- GRanges(seqnames = micro_feat_clean$chr,
                          ranges   = IRanges(as.numeric(micro_feat_clean$start), 
                                             as.numeric(micro_feat_clean$end)),
                          strand   = "*")

# harmonize chromosome naming style with TxDb (e.g., "chr1" vs "1")
seqlevelsStyle(micro_feat_gr) <- seqlevelsStyle(TxDb.Hsapiens.UCSC.hg38.knownGene)

# annotate peaks
micro_feat_anno <- annotatePeak(micro_feat_gr, 
                                TxDb=TxDb.Hsapiens.UCSC.hg38.knownGene, 
                                tssRegion=c(-3000,3000), 
                                annoDb="org.Hs.eg.db")

# merge annotate to res df
micro_feat_clean_anno <- merge(micro_feat_clean,
                               as.data.frame(micro_feat_anno@anno) %>%
                                 mutate(chr = str_sub(seqnames, 4, 6)),
                               by = c("chr", "start", "end"))
```

```{r oligo_peak_anno}
# convert peaks to granges obj
oligo_feat_gr  <- GRanges(seqnames = oligo_feat_clean$chr,
                          ranges   = IRanges(as.numeric(oligo_feat_clean$start), 
                                             as.numeric(oligo_feat_clean$end)),
                          strand   = "*")

# harmonize chromosome naming style with TxDb (e.g., "chr1" vs "1")
seqlevelsStyle(oligo_feat_gr) <- seqlevelsStyle(TxDb.Hsapiens.UCSC.hg38.knownGene)

# annotate peaks
oligo_feat_anno <- annotatePeak(oligo_feat_gr, 
                                TxDb=TxDb.Hsapiens.UCSC.hg38.knownGene, 
                                tssRegion=c(-3000,3000), 
                                annoDb="org.Hs.eg.db")

# merge annotate to res df
oligo_feat_clean_anno <- merge(oligo_feat_clean,
                               as.data.frame(oligo_feat_anno@anno) %>%
                                 mutate(chr = str_sub(seqnames, 4, 6)),
                               by = c("chr", "start", "end"))
```

```{r neuro_peak_anno}
# convert peaks to granges obj
neuro_feat_gr  <- GRanges(seqnames = neuro_feat_clean$chr,
                          ranges   = IRanges(as.numeric(neuro_feat_clean$start), 
                                             as.numeric(neuro_feat_clean$end)),
                          strand   = "*")

# harmonize chromosome naming style with TxDb (e.g., "chr1" vs "1")
seqlevelsStyle(neuro_feat_gr) <- seqlevelsStyle(TxDb.Hsapiens.UCSC.hg38.knownGene)

# annotate peaks
neuro_feat_anno <- annotatePeak(neuro_feat_gr, 
                                TxDb=TxDb.Hsapiens.UCSC.hg38.knownGene, 
                                tssRegion=c(-3000,3000), 
                                annoDb="org.Hs.eg.db")

# merge annotate to res df
neuro_feat_clean_anno <- merge(neuro_feat_clean,
                               as.data.frame(neuro_feat_anno@anno) %>%
                                 mutate(chr = str_sub(seqnames, 4, 6)),
                               by = c("chr", "start", "end"))
```

## Covariate distributions

```{r covariate_dist}
# sex by pd
sex_plot <- ggplot(neuro_meta, 
                   aes(x=group, fill = sex)) +
  geom_bar() +
  theme_bw()
sex_plot

# age by pd
age_plot <- ggplot(neuro_meta, 
                   aes(x=group,
                       y = age, 
                       fill = group)) +
  geom_boxplot() +
  geom_point() +
  theme_bw()
age_plot

age_ttest <- t.test(age ~ group, data=neuro_meta)
age_ttest # p = 0.03

# age by sex by pd
agesex_plot <- ggplot(neuro_meta, 
                      aes(x=sex,
                          y = age, 
                          fill = sex)) +
  geom_boxplot() +
  geom_point() +
  theme_bw()
agesex_plot

agesex_ttest <- t.test(age ~ sex, data=neuro_meta)
agesex_ttest # p = 0.07064

# frip by pd
frip_plot <- ggplot(neuro_meta, 
                    aes(x=group,
                        y = frip, 
                        fill = group)) +
  geom_boxplot() +
  geom_point() +
  theme_bw()
frip_plot

frip_ttest <- t.test(frip ~ group, data=neuro_meta)
frip_ttest # p = 0.27

# pmid by pd
pmi_plot <- ggplot(neuro_meta, 
                   aes(x=group,
                       y = pmi, 
                       fill = group)) +
  geom_boxplot() +
  geom_point() +
  theme_bw()
pmi_plot

pmi_ttest <- t.test(pmi ~ group, data=neuro_meta)
pmi_ttest # p = 0.66
```

```{r covariate_corr}
# convert all var to num for corr
meta_corr <- meta %>%
  dplyr::select(-c(sample, pid, celltype)) %>%
  mutate(across(where(is.character), as.factor)) %>%
  mutate(across(where(is.factor), as.numeric))

corr <- round(cor(meta_corr),2)
corr_long <- reshape2::melt(corr)

ggplot(data = corr_long, 
       aes(x=Var1, 
           y=Var2, 
           fill=value)) + 
  geom_tile(color = "white")+
  scale_fill_gradient2(low = "blue",
                       high = "red",
                       mid = "white", 
                       midpoint = 0, 
                       limit = c(-1,1), 
                       space = "Lab", 
                       name="Pearson\nCorrelation") +
  theme_minimal() +
  geom_text(aes(Var2, 
                Var1,
                label = as.character(value)), 
            color = "black",
            size = 4) 
```

## Differential analysis

### Differential analysis of microglia

```{r micro_diff_filter}

micro_meta$pmi_centered <- scale(micro_meta$pmi, scale = F, center = T)
# remove sex chromosomes
micro_feat_clean_noxy <- micro_feat_clean_anno %>%
  subset(!chr %in% c("X", "Y"))

# separate coordinates and counts
micro_peak_coords <- micro_feat_clean_noxy[, c(1:3,13:27)]
micro_counts <- as.matrix(micro_feat_clean_noxy[, -c(1:3,13:27)])

# recode group as factor
# set control as base level
micro_meta <- micro_meta %>%
  mutate(group = relevel(factor(group), 
                         ref = "control")) 

# create edgeR dgelist for filtering
micro_dgelist <- DGEList(counts = micro_counts, 
                         genes = micro_peak_coords,
                         samples = micro_meta)

# replicate-aware filtering - min count 5 
micro_keep <- filterByExpr(micro_dgelist, 
                           group = micro_dgelist$samples$group, 
                           min.count = 5,
                           min.total.count = 50) 

sum(micro_keep) # 90486

# remove those with <5 from count and peaks
micro_peak_coords_clean <- micro_peak_coords[micro_keep, ]
micro_counts_clean <- micro_counts[micro_keep, ]
```

```{r micro_diff_deseq2}

# make deseq2 object
# levels of h3k27ac decrease as soon as 24 PMI (https://pmc.ncbi.nlm.nih.gov/articles/PMC6330433/)
micro_dds = DESeqDataSetFromMatrix(countData = micro_counts_clean,
                                   colData = micro_meta,
                                   design = ~ group + sex + pmi_centered)

# diff analysis using deseq2
micro_dds_diff = DESeq(micro_dds, 
                       test = "Wald", 
                       fitType = "local",
                       # reduced = ~ sv1, # lrt only
                       sfType = "poscounts")

# get normalized peak counts (uses method from DESeq)
micro_dds_norm_counts = counts(micro_dds_diff, 
                               normalized = TRUE)  

# create results dataframe
micro_dds_res = results(micro_dds_diff, 
                        pAdjustMethod = "BH",
                        independentFiltering = FALSE, 
                        altHypothesis = "greaterAbs")

# get result names
# resultsNames(micro_dds_diff)

# # fold change shrinkage
micro_dds_res_shr <- lfcShrink(micro_dds_diff, 
                                coef = "group_pd_vs_control", 
                               type = "apeglm")

# make results df
micro_dds_res_df <- data.frame(baseMean = micro_dds_res$baseMean,
                               lfc = micro_dds_res$log2FoldChange,
                               lfcSE = micro_dds_res$lfcSE,
                               lfc_shr = micro_dds_res_shr$log2FoldChange,
                               stat = micro_dds_res$stat,
                               pvalue = micro_dds_res$pvalue,
                               padj = micro_dds_res$padj,
                               micro_peak_coords_clean) 

# get only sig hyper acetylated peaks
micro_dds_res_df_hyper <- micro_dds_res_df %>%
  subset(padj < 0.05 & lfc > 0) %>%
  subset((lfc > 0 & lfc_shr > 0))

dim(micro_dds_res_df_hyper)

# get only sig hypoacetylated peaks
micro_dds_res_df_hypo <- micro_dds_res_df %>%
  subset(padj < 0.05 & lfc < 0) %>%
  subset((lfc < 0 & lfc_shr <0))

nrow(micro_dds_res_df_hypo)
```

```{r micro_qqplot}
qq_p_gg(micro_dds_res_df$pvalue)
lambda_from_p(micro_dds_res_df$pvalue)
```

```{r micro_volcano}
ggplot(micro_dds_res_df, 
       aes(x = lfc, 
           y = -log(pvalue))) +
  geom_point() +
  theme_bw() 
```

```{r vis_top_peak_hyper}
# region (use NCBI-style chr names for hg38)
gen   <- "hg38"
chrom <- 11
start <- 31792734
end   <- 31819119

ax <- GenomeAxisTrack()

mart <- useEnsembl(biomart = "genes", 
                   dataset = "hsapiens_gene_ensembl")

grt <- BiomartGeneRegionTrack(
  chromosome= chrom,
  start = start,
  end = end,
  genome = gen,
  name        = "Ensembl (HGNC symbols)",
  biomart     = mart
)

rg <- ranges(grt)
sym <- mcols(rg)$symbol
gid <- mcols(rg)$gene
sym[is.na(sym) | sym == ""] <- gid[is.na(sym) | sym == ""]
mcols(rg)$symbol <- sym
grt <- GeneRegionTrack(rg, genome = gen, chromosome = chrom, name = "Genes")


displayPars(grt) <- list(
  transcriptAnnotation = "symbol",  # show HGNC symbols
  geneSymbol = TRUE,
  collapseTranscripts = "meta",     # "meta" | "longest" | "gene" | FALSE
  fontsize = 10
)

# ---- signal tracks (BigWig) ----
# Tip: You can also use type="h" (horizon), "histogram", or "p" (points). Histogram is nice for CPM.
control_tr <- DataTrack(
  range       = ctl_bw,
  genome      = gen,
  chromosome  = chrom,
  type        = "histogram",
  col.histogram="darkblue",
  fill.histogram="darkblue",
  name        = "Control (CPM)",
)

pd_tr <- DataTrack(
  range       = dis_bw,
  genome      = gen,
  chromosome  = chrom,
  type        = "histogram",
  col.histogram="darkred", 
  fill.histogram="darkred",
  name        = "Disease (CPM)"
)


# --- define the differential peak span ---
peak_start <- 31802734
peak_end   <- 31809119


# Build a clean GRanges with an 'id' column (character, length == n features)
diff_peak <- GRanges(seqnames = chrom, ranges = IRanges(peak_start, peak_end))
mcols(diff_peak)$id <- NA  # label to display

# Sanity check: id length matches feature count
stopifnot(length(mcols(diff_peak)$id) == length(diff_peak))

# Make the boxed AnnotationTrack
diff_tr <- AnnotationTrack(
  range      = diff_peak,
  genome     = gen,
  chromosome = chrom,
  col        = "black",
  fill = "black",
  lwd        = 2,
  stacking   = "dense"
)


displayPars(diff_tr) <- c(
  displayPars(diff_tr),
  list(
    featureAnnotation = NULL,  # draw text from mcols(...)$id
    fontsize          = 9,
    just              = "above"  # place label above feature
  )
)

# 4) Plot (put genes between axis and signal if you like)
plotTracks(list(ax, grt, pd_tr, control_tr, diff_tr),
           chromosome = chrom, from = start, to = end,
           background.title = "white", col.title = "black")
```



```{r vis_top_peak_hypo}
# region (use NCBI-style chr names for hg38)
gen   <- "hg38"
chrom <- 11
start <- 31792734
end   <- 31819119

ax <- GenomeAxisTrack()

mart <- useEnsembl(biomart = "genes", 
                   dataset = "hsapiens_gene_ensembl")

grt <- BiomartGeneRegionTrack(
  chromosome= chrom,
  start = start,
  end = end,
  genome = gen,
  name        = "Ensembl (HGNC symbols)",
  biomart     = mart
)

rg <- ranges(grt)
sym <- mcols(rg)$symbol
gid <- mcols(rg)$gene
sym[is.na(sym) | sym == ""] <- gid[is.na(sym) | sym == ""]
mcols(rg)$symbol <- sym
grt <- GeneRegionTrack(rg, genome = gen, chromosome = chrom, name = "Genes")


displayPars(grt) <- list(
  transcriptAnnotation = "symbol",  # show HGNC symbols
  geneSymbol = TRUE,
  collapseTranscripts = "meta",     # "meta" | "longest" | "gene" | FALSE
  fontsize = 10
)

# ---- signal tracks (BigWig) ----
# Tip: You can also use type="h" (horizon), "histogram", or "p" (points). Histogram is nice for CPM.
control_tr <- DataTrack(
  range       = ctl_bw,
  genome      = gen,
  chromosome  = chrom,
  type        = "histogram",
  col.histogram="darkblue",
  fill.histogram="darkblue",
  name        = "Control (CPM)",
)

pd_tr <- DataTrack(
  range       = dis_bw,
  genome      = gen,
  chromosome  = chrom,
  type        = "histogram",
  col.histogram="darkred", 
  fill.histogram="darkred",
  name        = "Disease (CPM)"
)


# --- define the differential peak span ---
peak_start <- 31802734
peak_end   <- 31809119


# Build a clean GRanges with an 'id' column (character, length == n features)
diff_peak <- GRanges(seqnames = chrom, ranges = IRanges(peak_start, peak_end))
mcols(diff_peak)$id <- NA  # label to display

# Sanity check: id length matches feature count
stopifnot(length(mcols(diff_peak)$id) == length(diff_peak))

# Make the boxed AnnotationTrack
diff_tr <- AnnotationTrack(
  range      = diff_peak,
  genome     = gen,
  chromosome = chrom,
  col        = "black",
  fill = "black",
  lwd        = 2,
  stacking   = "dense"
)


displayPars(diff_tr) <- c(
  displayPars(diff_tr),
  list(
    featureAnnotation = NULL,  # draw text from mcols(...)$id
    fontsize          = 9,
    just              = "above"  # place label above feature
  )
)

# 4) Plot (put genes between axis and signal if you like)
plotTracks(list(ax, grt, pd_tr, control_tr, diff_tr),
           chromosome = chrom, from = start, to = end,
           background.title = "white", col.title = "black")
```


```{r micro_go_terms_hyper}
# convert peaks to granges obj
micro_dds_res_df_hyper_gr <- GRanges(seqnames = micro_dds_res_df_hyper$chr,
                                     ranges   = IRanges(as.numeric(micro_dds_res_df_hyper$start), 
                                                        as.numeric(micro_dds_res_df_hyper$end)),
                                     strand   = "*")

# harmonize chromosome naming style with TxDb (e.g., "chr1" vs "1")
seqlevelsStyle(micro_dds_res_df_hyper_gr) <- seqlevelsStyle(TxDb.Hsapiens.UCSC.hg38.knownGene)

# annotate peaks
micro_dds_res_df_hyper_anno <- annotatePeak(micro_dds_res_df_hyper_gr, 
                                            TxDb=TxDb.Hsapiens.UCSC.hg38.knownGene, 
                                            tssRegion=c(-3000,3000), 
                                            annoDb="org.Hs.eg.db")

# merge annotate to res df
micro_dds_res_df_hyper <- merge(micro_dds_res_df_hyper,
                                as.data.frame(micro_dds_res_df_hyper_anno@anno) %>%
                                  mutate(chr = str_sub(seqnames, 4, 6)),
                                by = c("chr", "start", "end"))

micro_hyper_genes <- unique(na.omit(as.data.frame(micro_dds_res_df_hyper_anno)$ENSEMBL))

micro_hyper_go <- enrichGO(gene=micro_hyper_genes, 
                           OrgDb=org.Hs.eg.db, 
                           keyType="ENSEMBL",
                           ont="BP", 
                           #pvalueCutoff=0.01, 
                           qvalueCutoff=0.05)
dotplot(micro_hyper_go)
```

```{r micro_go_terms_hypo}
# convert peaks to granges obj
micro_dds_res_df_hypo_gr <- GRanges(seqnames = micro_dds_res_df_hypo$chr,
                                    ranges   = IRanges(as.numeric(micro_dds_res_df_hypo$start), 
                                                       as.numeric(micro_dds_res_df_hypo$end)),
                                    strand   = "*")

# harmonize chromosome naming style with TxDb (e.g., "chr1" vs "1")
seqlevelsStyle(micro_dds_res_df_hypo_gr) <- seqlevelsStyle(TxDb.Hsapiens.UCSC.hg38.knownGene)

# annotate peaks
micro_dds_res_df_hypo_anno <- annotatePeak(micro_dds_res_df_hypo_gr, 
                                           TxDb=TxDb.Hsapiens.UCSC.hg38.knownGene, 
                                           tssRegion=c(-3000,3000), 
                                           annoDb="org.Hs.eg.db")

# merge annotate to res df
micro_dds_res_df_hypo <- merge(micro_dds_res_df_hypo,
                               as.data.frame(micro_dds_res_df_hypo_anno@anno) %>%
                                 mutate(chr = str_sub(seqnames, 4, 6)),
                               by = c("chr", "start", "end"))

micro_hypo_genes <- unique(na.omit(as.data.frame(micro_dds_res_df_hypo_anno)$ENSEMBL))

micro_hypo_go <- enrichGO(gene=micro_hypo_genes, 
                          OrgDb=org.Hs.eg.db, 
                          keyType="ENSEMBL",
                          ont="BP", 
                          #pvalueCutoff=0.01, 
                          qvalueCutoff=0.05)
dotplot(micro_hypo_go)
```



### Differential analysis of oligodendrocytes

```{r oligo_diff_filter}
# remove sex chromosomes
oligo_feat_clean_noxy <- oligo_feat_clean_anno %>%
  subset(!chr %in% c("X", "Y"))

# separate coordinates and counts 
oligo_peak_coords <- oligo_feat_clean_noxy[, c(1:3, 13:27)]
oligo_counts <- as.matrix(oligo_feat_clean_noxy[, -c(1:3, 13:27)])

# recode group as factor
# set control as base level
oligo_meta <- oligo_meta %>%
  mutate(group = relevel(factor(group), 
                         ref = "control")) 

# create edgeR dgelist for filtering
oligo_dgelist <- DGEList(counts = oligo_counts, 
                         samples = oligo_meta)

# replicate-aware filtering - min count 5 
oligo_keep <- filterByExpr(oligo_dgelist, 
                           group = oligo_dgelist$samples$group, 
                           min.count = 5, 
                           min.total.count=50) # smallest group (control) has 4 samples

sum(oligo_keep) # 76711

# remove those with <5 from count and peaks
oligo_peak_coords_clean <- oligo_peak_coords[oligo_keep, ]
oligo_counts_clean <- oligo_counts[oligo_keep, ]
```

```{r oligo_diff_deseq2}
# make deseq2 object
oligo_dds = DESeqDataSetFromMatrix(countData = oligo_counts_clean,
                                   colData = oligo_meta,
                                   design = ~ group + sex + pmi)

# diff analysis using deseq2
oligo_dds_diff = DESeq(oligo_dds, 
                       test = "Wald", 
                       fitType = "local",
                       # reduced = ~ sv1, # lrt only
                       sfType = "poscounts")

# get normalized peak counts (uses method from DESeq)
oligo_dds_norm_counts = counts(oligo_dds_diff, 
                               normalized = TRUE)  

# create results dataframe
oligo_dds_res = results(oligo_dds_diff, 
                        pAdjustMethod = "BH",
                        independentFiltering = FALSE, 
                        altHypothesis = "greaterAbs")

# get result names
# resultsNames(oligo_dds_diff)

# # fold change shrinkage
# oligo_dds_res_shr <- lfcShrink(oligo_dds_diff, 
#                                coef = "group_pd_vs_control", 
#                                type = "apeglm")

# make results df
oligo_dds_res_df <- data.frame(chr = oligo_peak_coords_clean$chr,
                               start = oligo_peak_coords_clean$start,
                               end = oligo_peak_coords_clean$end,
                               baseMean = oligo_dds_res$baseMean,
                               lfc = oligo_dds_res$log2FoldChange,
                               lfcSE = oligo_dds_res$lfcSE,
                               # lfc_shr = oligo_dds_res_shr$log2FoldChange,
                               stat = oligo_dds_res$stat,
                               pvalue = oligo_dds_res$pvalue,
                               padj = oligo_dds_res$padj) 

# get only sig hyper acetylated peaks
oligo_dds_res_df_hyper <- oligo_dds_res_df %>%
  subset(padj < 0.05 & lfc > 0)

dim(oligo_dds_res_df_hyper)

# get only sig hypoacetylated peaks
oligo_dds_res_df_hypo <- oligo_dds_res_df %>%
  subset(padj < 0.05 & lfc < 0)

nrow(oligo_dds_res_df_hypo)
```


```{r oligo_qqplot}
qq_p_gg(oligo_dds_res_df$pvalue)
lambda_from_p(oligo_dds_res_df$pvalue)
```

```{r oligo_volcano}
ggplot(oligo_dds_res_df, 
       aes(x = lfc, 
           y = -log(pvalue))) +
  geom_point() +
  theme_bw() 
```

```{r oligo_go_terms_hyper}
# convert peaks to granges obj
oligo_dds_res_df_hyper_gr <- GRanges(seqnames = oligo_dds_res_df_hyper$chr,
                                     ranges   = IRanges(as.numeric(oligo_dds_res_df_hyper$start), 
                                                        as.numeric(oligo_dds_res_df_hyper$end)),
                                     strand   = "*")

# harmonize chromosome naming style with TxDb (e.g., "chr1" vs "1")
seqlevelsStyle(oligo_dds_res_df_hyper_gr) <- seqlevelsStyle(TxDb.Hsapiens.UCSC.hg38.knownGene)

# annotate peaks
oligo_dds_res_df_hyper_anno <- annotatePeak(oligo_dds_res_df_hyper_gr, 
                                            TxDb=TxDb.Hsapiens.UCSC.hg38.knownGene, 
                                            tssRegion=c(-3000,3000), 
                                            annoDb="org.Hs.eg.db")

# merge annotate to res df
oligo_dds_res_df_hyper <- merge(oligo_dds_res_df_hyper,
                                as.data.frame(oligo_dds_res_df_hyper_anno@anno) %>%
                                  mutate(chr = str_sub(seqnames, 4, 6)),
                                by = c("chr", "start", "end"))

oligo_hyper_genes <- unique(na.omit(as.data.frame(oligo_dds_res_df_hyper_anno)$ENSEMBL))

oligo_hyper_go <- enrichGO(gene=oligo_hyper_genes, 
                           OrgDb=org.Hs.eg.db, 
                           keyType="ENSEMBL",
                           ont="BP", 
                           #pvalueCutoff=0.01, 
                           qvalueCutoff=0.05)
dotplot(oligo_hyper_go)
```

```{r oligo_go_terms_hypo}
# convert peaks to granges obj
oligo_dds_res_df_hypo_gr <- GRanges(seqnames = oligo_dds_res_df_hypo$chr,
                                    ranges   = IRanges(as.numeric(oligo_dds_res_df_hypo$start), 
                                                       as.numeric(oligo_dds_res_df_hypo$end)),
                                    strand   = "*")

# harmonize chromosome naming style with TxDb (e.g., "chr1" vs "1")
seqlevelsStyle(oligo_dds_res_df_hypo_gr) <- seqlevelsStyle(TxDb.Hsapiens.UCSC.hg38.knownGene)

# annotate peaks
oligo_dds_res_df_hypo_anno <- annotatePeak(oligo_dds_res_df_hypo_gr, 
                                           TxDb=TxDb.Hsapiens.UCSC.hg38.knownGene, 
                                           tssRegion=c(-3000,3000), 
                                           annoDb="org.Hs.eg.db")

# merge annotate to res df
oligo_dds_res_df_hypo <- merge(oligo_dds_res_df_hypo,
                               as.data.frame(oligo_dds_res_df_hypo_anno@anno) %>%
                                 mutate(chr = str_sub(seqnames, 4, 6)),
                               by = c("chr", "start", "end"))

oligo_hypo_genes <- unique(na.omit(as.data.frame(oligo_dds_res_df_hypo_anno)$ENSEMBL))

oligo_hypo_go <- enrichGO(gene=oligo_hypo_genes, 
                          OrgDb=org.Hs.eg.db, 
                          keyType="ENSEMBL",
                          ont="BP", 
                          #pvalueCutoff=0.01, 
                          qvalueCutoff=0.05)
dotplot(oligo_hypo_go)
```


### Differential analysis of neuro

```{r neuro_diff_filter}
# remove sex chromosomes
neuro_feat_clean_noxy <- neuro_feat_clean %>%
  subset(!chr %in% c("X", "Y"))

# separate coordinates and counts
neuro_peak_coords <- neuro_feat_clean[, 1:3]
neuro_counts <- as.matrix(neuro_feat_clean[, -(1:3)])

# recode group as factor
# set control as base level
neuro_meta <- neuro_meta %>%
  mutate(group = relevel(factor(group), 
                         ref = "control")) 

# create edgeR dgelist for filtering
neuro_dgelist <- DGEList(counts = neuro_counts, 
                         samples = neuro_meta)

# replicate-aware filtering - min count 5 
neuro_keep <- filterByExpr(neuro_dgelist, 
                           group = neuro_dgelist$samples$group, 
                           min.count = 5,
                           min.total.count=50) # smallest group (control) has 4 samples

# remove those with <5 from count and peaks
neuro_peak_coords_clean <- neuro_peak_coords[neuro_keep, ]
neuro_counts_clean <- neuro_counts[neuro_keep, ]

# recreate neuro dgelist
neuro_dgelist <- DGEList(counts = neuro_counts_clean, 
                         genes = neuro_peak_coords_clean,
                         samples = neuro_meta)
```


```{r neuro_diff_deseq2}
# make deseq2 object
neuro_dds = DESeqDataSetFromMatrix(countData = neuro_counts_clean,
                                   colData = neuro_meta,
                                   design = ~ group + sex + pmi)

# diff analysis using deseq2
neuro_dds_diff = DESeq(neuro_dds, 
                       test = "Wald", 
                       fitType = "local",
                       # reduced = ~ sv1, # lrt only
                       sfType = "poscounts")

# get normalized peak counts (uses method from DESeq)
neuro_dds_norm_counts = counts(neuro_dds_diff, 
                               normalized = TRUE)  

# create results dataframe
neuro_dds_res = results(neuro_dds_diff, 
                        pAdjustMethod = "BH",
                        independentFiltering = FALSE, 
                        altHypothesis = "greaterAbs")

# get result names
# resultsNames(neuro_dds_diff)

# # fold change shrinkage
# neuro_dds_res_shr <- lfcShrink(neuro_dds_diff, 
#                                coef = "group_pd_vs_control", 
#                                type = "apeglm")

# make results df
neuro_dds_res_df <- data.frame(chr = neuro_peak_coords_clean$chr,
                               start = neuro_peak_coords_clean$start,
                               end = neuro_peak_coords_clean$end,
                               baseMean = neuro_dds_res$baseMean,
                               lfc = neuro_dds_res$log2FoldChange,
                               lfcSE = neuro_dds_res$lfcSE,
                               # lfc_shr = neuro_dds_res_shr$log2FoldChange,
                               stat = neuro_dds_res$stat,
                               pvalue = neuro_dds_res$pvalue,
                               padj = neuro_dds_res$padj) 

```


```{r neuro_qqplot}
qq_p_gg(neuro_dds_res_df$pvalue)
lambda_from_p(neuro_dds_res_df$pvalue)
```

```{r neuro_volcano}
ggplot(neuro_dds_res_df, 
       aes(x = lfc, 
           y = -log(pvalue))) +
  geom_point() +
  theme_bw() 
```





### Output size factors

```{r size factors}
# get size factors
micro_size_factors <- sizeFactors(micro_dds_diff) %>%
  as.data.frame() %>%
  rownames_to_column() %>%
  merge(.,
        micro_meta %>% 
          dplyr::select(sample, celltype, group),
        by.x = "rowname", by.y= "sample")

oligo_size_factors <- sizeFactors(oligo_dds_diff) %>%
  as.data.frame() %>%
  rownames_to_column() %>%
  merge(.,
        oligo_meta %>% 
          dplyr::select(sample, celltype, group),
        by.x = "rowname", by.y= "sample")

neuro_size_factors <- sizeFactors(neuro_dds_diff) %>%
  as.data.frame() %>%
  rownames_to_column() %>%
  merge(.,
        neuro_meta %>% 
          dplyr::select(sample, celltype, group),
        by.x = "rowname", by.y= "sample")

size_factors <- rbind(micro_size_factors,
                      oligo_size_factors,
                      neuro_size_factors) %>%
  as.data.frame() 

write_tsv(size_factors,
          here("data_out",
               "080_differential_analysis",
               "size_factors.tsv"),
          col_names = FALSE)
```


