---
title: "Differential analysis of bulk CUT&Tag data"
author: "Samantha Lee"
date: "`r Sys.Date()`"
output:
  html_document:
    code_folding: hide
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE,
                      message = FALSE,
                      warning = FALSE)
```

## Load data 

Load required libraries
```{r req_libs}
library(here) # dir calling
library(tidyverse) # data wrangling
library(sva) # surrogate variables
library(DESeq2) # diff analysis
library(edgeR) # diff analysis 
library(ChIPseeker) # peak anno
library(clusterProfiler) # go analysis
library(TxDb.Hsapiens.UCSC.hg38.knownGene) # anno library
library(org.Hs.eg.db) # human annotation
library(Gviz) # ucsc track vis
library(GenomicRanges) # gr objects
library(rtracklayer) # import data
library(biomaRt) # get annotations
library(car) # vif
library(bacon) # bayesian adjustment for genomic inflation 
```

Source required custom function
```{r src}
# qqplot and lambda
source(here("r",
            "qqplot.r"))
```

Read in toptables from age analysis of healthy AD controls (from Paulina) and subset statistically significant (BH p < 0.05) hits. Convert to GRanges object. 
```{r data_in_age_tt}
micro_age <- read.csv(here("resources",
                           "age_assoc_peaks",
                           "IRF8_mergebam_peaks_all_deouts_ageContinuous.csv"))

micro_age_sig <- micro_age %>%
  subset(padj_BH < 0.05) %>%
  mutate(chr = str_split_i(peakName, pattern = "[.]", i=1),
         start = str_split_i(peakName, pattern = "[.]", i=2),
         end = str_split_i(peakName, pattern = "[.]", i=3)) %>% 
  dplyr::select(-peakName)

micro_age_sig_gr <-  makeGRangesFromDataFrame(micro_age_sig,
                                              keep.extra.columns = TRUE,
                                              seqnames.field = "chr",
                                              start.field = "start",
                                              end.field = "end")

oligo_age <- read.csv(here("resources",
                           "age_assoc_peaks",
                           "SOX10_mergebam_peaks_all_deouts_ageContinuous.csv"))

oligo_age_sig <- oligo_age %>%
  subset(padj_BH < 0.05) %>%
  mutate(chr = str_split_i(peakName, pattern = "[.]", i=1),
         start = str_split_i(peakName, pattern = "[.]", i=2),
         end = str_split_i(peakName, pattern = "[.]", i=3)) %>% 
  dplyr::select(-peakName)

oligo_age_sig_gr <-  makeGRangesFromDataFrame(oligo_age_sig,
                                              keep.extra.columns = TRUE,
                                              seqnames.field = "chr",
                                              start.field = "start",
                                              end.field = "end")

neuro_age <- read.csv(here("resources",
                           "age_assoc_peaks",
                           "NeuN_mergebam_peaks_all_deouts_ageContinuous.csv"))

neuro_age_sig <- neuro_age %>%
  subset(padj_BH < 0.05) %>%
  mutate(chr = str_split_i(peakName, pattern = "[.]", i=1),
         start = str_split_i(peakName, pattern = "[.]", i=2),
         end = str_split_i(peakName, pattern = "[.]", i=3)) %>% 
  dplyr::select(-peakName)

neuro_age_sig_gr <-  makeGRangesFromDataFrame(neuro_age_sig,
                                              keep.extra.columns = TRUE,
                                              seqnames.field = "chr",
                                              start.field = "start",
                                              end.field = "end")
```

Read in participant metadata.
```{r data_in_meta}
meta <- read.csv(here("resources",
                      "metadata",
                      "metadata_participants.csv")) %>%
  mutate(group = ifelse(grepl("PDC", pid), "control", "pd")) %>%
  mutate(sample = str_extract(sample, "IGF\\d+"))
```

Read in FRiP scores. 
```{r data_in_frip_scores}
frip_in <- function(f) {
  read_table(f,
             skip = 0, 
             col_names = T)
}

frip_scores <- map_dfr(list.files(here("data_out",
                                       "070_peaks", 
                                       "072_frip_scores"),
                                  pattern = "^.*\\.txt$",
                                  full.names = TRUE), 
                       frip_in) 
```

Combine metadata with FRiP score.
```{r combine_meta_frip}
meta <- merge(meta,
              frip_scores %>%
                dplyr::select(file, percent) %>%
                dplyr::rename("frip" = percent,
                              "sample" = file),
              by = "sample")
```

Read in celltype-specific feature counts.
```{r data_in_feature_counts}
micro_tsv_files <- list.files(here("data_out",
                                   "070_peaks",
                                   "078_peaks_feature_counts_celltype"), 
                              pattern = "*microglia*", 
                              full.names = TRUE)

oligo_tsv_files <- list.files(here("data_out",
                                   "070_peaks",
                                   "078_peaks_feature_counts_celltype"), 
                              pattern = "*oligodendrocyte*", 
                              full.names = TRUE)

neuro_tsv_files <- list.files(here("data_out",
                                   "070_peaks",
                                   "078_peaks_feature_counts_celltype"), 
                              pattern = "*neuron*", 
                              full.names = TRUE)

micro_feat <- purrr::imap_dfc(
  micro_tsv_files,
  ~ read_tsv(.x, show_col_types = FALSE, col_names = FALSE) %>% 
    { if (.y == 1) dplyr::select(., 1:3, dplyr::last_col())
      else          dplyr::select(., dplyr::last_col()) }
)


oligo_feat <- purrr::imap_dfc(
  oligo_tsv_files,
  ~ read_tsv(.x, show_col_types = FALSE, col_names = FALSE) %>% 
    { if (.y == 1) dplyr::select(., 1:3, dplyr::last_col())
      else          dplyr::select(., dplyr::last_col()) }
)

neuro_feat <- purrr::imap_dfc(
  neuro_tsv_files,
  ~ read_tsv(.x, show_col_types = FALSE, col_names = FALSE) %>% 
    { if (.y == 1) dplyr::select(., 1:3, dplyr::last_col())
      else          dplyr::select(., dplyr::last_col()) }
)

colnames(micro_feat) <- c("chr", "start", "end",
                          sub("_microglia_multicov\\.tsv$", 
                              "", 
                              basename(micro_tsv_files)))

colnames(oligo_feat) <- c("chr", "start", "end",
                          sub("_oligodendrocytes_multicov\\.tsv$", 
                              "", 
                              basename(oligo_tsv_files)))

colnames(neuro_feat) <- c("chr", "start", "end",
                          sub("_neuron_multicov\\.tsv$", 
                              "", 
                              basename(neuro_tsv_files)))

# remove low frip samples
micro_feat_clean <- micro_feat %>%
  dplyr::select(!"IGF136817")

oligo_feat_clean <- oligo_feat %>%
  dplyr::select(!"IGF136816")

neuro_feat_clean <- neuro_feat %>%
  dplyr::select(!"IGF136815")

# remove uneeded files 
rm(micro_tsv_files,
   oligo_tsv_files,
   neuro_tsv_files,
   micro_feat,
   oligo_feat,
   neuro_feat,
   frip_scores)
```

Subset metadata for each cell type
```{r meta_celltype}
micro_meta <- meta %>%
  mutate(sample = str_extract(sample, "IGF\\d+")) %>%
  subset(sample %in% colnames(micro_feat_clean))

oligo_meta <- meta %>%
  mutate(sample = str_extract(sample, "IGF\\d+")) %>%
  subset(sample %in% colnames(oligo_feat_clean))

neuro_meta <- meta %>%
  mutate(sample = str_extract(sample, "IGF\\d+")) %>%
  subset(sample %in% colnames(neuro_feat_clean))
```

## Annotate peaks for each cell type

```{r micro_peak_anno}
# convert peaks to granges obj
micro_feat_gr  <- GRanges(seqnames = micro_feat_clean$chr,
                          ranges   = IRanges(as.numeric(micro_feat_clean$start), 
                                             as.numeric(micro_feat_clean$end)),
                          strand   = "*")

# harmonize chromosome naming style with TxDb (e.g., "chr1" vs "1")
seqlevelsStyle(micro_feat_gr) <- seqlevelsStyle(TxDb.Hsapiens.UCSC.hg38.knownGene)

# annotate peaks
micro_feat_anno <- annotatePeak(micro_feat_gr, 
                                TxDb=TxDb.Hsapiens.UCSC.hg38.knownGene, 
                                tssRegion=c(-3000,3000), 
                                annoDb="org.Hs.eg.db")

# merge annotate to res df
micro_feat_clean_anno <- merge(micro_feat_clean,
                               as.data.frame(micro_feat_anno@anno) %>%
                                 mutate(chr = str_sub(seqnames, 4, 6)),
                               by = c("chr", "start", "end"))
```

```{r oligo_peak_anno}
# convert peaks to granges obj
oligo_feat_gr  <- GRanges(seqnames = oligo_feat_clean$chr,
                          ranges   = IRanges(as.numeric(oligo_feat_clean$start), 
                                             as.numeric(oligo_feat_clean$end)),
                          strand   = "*")

# harmonize chromosome naming style with TxDb (e.g., "chr1" vs "1")
seqlevelsStyle(oligo_feat_gr) <- seqlevelsStyle(TxDb.Hsapiens.UCSC.hg38.knownGene)

# annotate peaks
oligo_feat_anno <- annotatePeak(oligo_feat_gr, 
                                TxDb=TxDb.Hsapiens.UCSC.hg38.knownGene, 
                                tssRegion=c(-3000,3000), 
                                annoDb="org.Hs.eg.db")

# merge annotate to res df
oligo_feat_clean_anno <- merge(oligo_feat_clean,
                               as.data.frame(oligo_feat_anno@anno) %>%
                                 mutate(chr = str_sub(seqnames, 4, 6)),
                               by = c("chr", "start", "end"))
```

```{r neuro_peak_anno}
# convert peaks to granges obj
neuro_feat_gr  <- GRanges(seqnames = neuro_feat_clean$chr,
                          ranges   = IRanges(as.numeric(neuro_feat_clean$start), 
                                             as.numeric(neuro_feat_clean$end)),
                          strand   = "*")

# harmonize chromosome naming style with TxDb (e.g., "chr1" vs "1")
seqlevelsStyle(neuro_feat_gr) <- seqlevelsStyle(TxDb.Hsapiens.UCSC.hg38.knownGene)

# annotate peaks
neuro_feat_anno <- annotatePeak(neuro_feat_gr, 
                                TxDb=TxDb.Hsapiens.UCSC.hg38.knownGene, 
                                tssRegion=c(-3000,3000), 
                                annoDb="org.Hs.eg.db")

# merge annotate to res df
neuro_feat_clean_anno <- merge(neuro_feat_clean,
                               as.data.frame(neuro_feat_anno@anno) %>%
                                 mutate(chr = str_sub(seqnames, 4, 6)),
                               by = c("chr", "start", "end"))
```

## Covariate distributions

```{r covariate_dist}
# sex by pd
sex_plot <- ggplot(neuro_meta, 
                   aes(x=group, fill = sex)) +
  geom_bar() +
  theme_bw()
sex_plot

# age by pd
age_plot <- ggplot(neuro_meta, 
                   aes(x=group,
                       y = age, 
                       fill = group)) +
  geom_boxplot() +
  geom_point() +
  theme_bw()
age_plot

age_ttest <- t.test(age ~ group, data=neuro_meta)
age_ttest # p = 0.03

# age by sex by pd
agesex_plot <- ggplot(neuro_meta, 
                      aes(x=sex,
                          y = age, 
                          fill = sex)) +
  geom_boxplot() +
  geom_point() +
  theme_bw()
agesex_plot

agesex_ttest <- t.test(age ~ sex, data=neuro_meta)
agesex_ttest # p = 0.07064

# frip by pd
frip_plot <- ggplot(neuro_meta, 
                    aes(x=group,
                        y = frip, 
                        fill = group)) +
  geom_boxplot() +
  geom_point() +
  theme_bw()
frip_plot

frip_ttest <- t.test(frip ~ group, data=neuro_meta)
frip_ttest # p = 0.27

# pmid by pd
pmi_plot <- ggplot(neuro_meta, 
                   aes(x=group,
                       y = pmi, 
                       fill = group)) +
  geom_boxplot() +
  geom_point() +
  theme_bw()
pmi_plot

pmi_ttest <- t.test(pmi ~ group, data=neuro_meta)
pmi_ttest # p = 0.66
```

```{r covariate_corr}
# convert all var to num for corr
meta_corr <- meta %>%
  dplyr::select(-c(sample, pid, celltype)) %>%
  mutate(across(where(is.character), as.factor)) %>%
  mutate(across(where(is.factor), as.numeric))

corr <- round(cor(meta_corr),2)
corr_long <- reshape2::melt(corr)

ggplot(data = corr_long, 
       aes(x=Var1, 
           y=Var2, 
           fill=value)) + 
  geom_tile(color = "white")+
  scale_fill_gradient2(low = "blue",
                       high = "red",
                       mid = "white", 
                       midpoint = 0, 
                       limit = c(-1,1), 
                       space = "Lab", 
                       name="Pearson\nCorrelation") +
  theme_minimal() +
  geom_text(aes(Var2, 
                Var1,
                label = as.character(value)), 
            color = "black",
            size = 4) 
```

## Differential analysis

### Differential analysis of microglia

```{r micro_diff_filter}
# remove sex chromosomes
micro_feat_clean_noxy <- micro_feat_clean_anno %>%
  subset(!chr %in% c("X", "Y"))

# separate coordinates and counts
micro_peak_coords <- micro_feat_clean_noxy[, c(1:3,13:27)]
micro_counts <- as.matrix(micro_feat_clean_noxy[, -c(1:3,13:27)])

# recode group as factor
# set control as base level
micro_meta <- micro_meta %>%
  mutate(group = relevel(factor(group), 
                         ref = "control")) 

# create edgeR dgelist for filtering
micro_dgelist <- DGEList(counts = micro_counts, 
                         genes = micro_peak_coords,
                         samples = micro_meta)

# replicate-aware filtering - min count 5 
micro_keep <- filterByExpr(micro_dgelist, 
                           group = micro_dgelist$samples$group, 
                           min.count = 5,
                           min.total.count = 50) 

sum(micro_keep) # 90486

# remove those with <5 from count and peaks
micro_peak_coords_clean <- micro_peak_coords[micro_keep, ]
micro_counts_clean <- micro_counts[micro_keep, ]

# make deseq2 object
# levels of h3k27ac decrease as soon as 24 PMI (https://pmc.ncbi.nlm.nih.gov/articles/PMC6330433/)
micro_dds = DESeqDataSetFromMatrix(countData = micro_counts_clean,
                                   colData = micro_meta %>%
                                     mutate(pmi_c = scale(pmi, 
                                                          scale = F, 
                                                          center = T)),
                                   design = ~ group + pmi_c + sex)
```

Examine sample clustering
```{r micro_dds_sample_clustering}
colData(micro_dds)$condition <- colData(micro_dds)$group  # or whatever column
vsd <- vst(micro_dds)

plotPCA(vsd) # top 500  
plotPCA(vsd, n=90618)  # all (equivalent to below) 

mat <- assay(vsd)
pca <- prcomp(t(mat), scale. = FALSE)
pc <- as.data.frame(pca$x)
pc$sample <- rownames(pc)

# Euclidean distance from the multivariate center
center <- colMeans(pc[,1:5])   # using first 5 PCs is common
d <- sqrt(rowSums((pc[,1:5] - center)^2))

# Convert distances to Z-scores
z <- scale(d)

# Mark outliers (|Z| > 3)
pc$outlier <- abs(z) > 3
```

```{r micro_check_vif}
md <- as.data.frame(colData(micro_dds))

# Ensure factors are coded as factors; pmi numeric
md$pd  <- as.factor(md$group)
md$sex <- as.factor(md$sex)
md$pmi <- as.numeric(md$pmi)

# Fit a dummy linear model (response is random noise)
set.seed(1)
fit <- lm(rnorm(nrow(md)) ~ group + pmi_c + sex, data = md)

# VIFs (GVIF for multi-df factors)
vif(fit)
# group      pmi_      sex 
# 1.036150 1.607185 1.597796 
```

```{r micro_diff_deseq2}
# diff analysis using deseq2
micro_dds_diff = DESeq(micro_dds, 
                       test = "Wald", 
                       fitType = "local",
                       minReplicatesForReplace = 5,
                       # reduced = ~ sv1, # lrt only
                       sfType = "poscounts")

# get normalized peak counts (uses method from DESeq)
micro_dds_norm_counts = counts(micro_dds_diff, 
                               normalized = TRUE)  

# create results dataframe
micro_dds_res = results(micro_dds_diff, 
                        pAdjustMethod = "BH",
                        independentFiltering = FALSE, 
                        altHypothesis = "greaterAbs")

# get result names
# resultsNames(micro_dds_diff)
# 
# # # fold change shrinkage
# micro_dds_res_shr <- lfcShrink(micro_dds_diff, 
#                                 coef = "group_pd_vs_control", 
#                                type = "apeglm")

# make results df
micro_dds_res_df <- data.frame(baseMean = micro_dds_res$baseMean,
                               lfc = micro_dds_res$log2FoldChange,
                               lfcSE = micro_dds_res$lfcSE,
                               #lfc_shr = micro_dds_res_shr$log2FoldChange,
                               stat = micro_dds_res$stat,
                               pvalue = micro_dds_res$pvalue,
                               padj = micro_dds_res$padj,
                               micro_peak_coords_clean, 
                               micro_dds_norm_counts) 

bc <- bacon(micro_dds_res_df$stat)
micro_dds_res_df$padj_bacon <- pval(bc) %>% p.adjust(method = "BH")
micro_dds_res_df$pvalue_bacon <- pval(bc)

# get normalized counts after accounting for covariates
vsd <- vst(micro_dds)

# Adjust for covariates batch + sex
assay_adj <- limma::removeBatchEffect(
   # assay(vsd),
  micro_dds_norm_counts,
    covariates = model.matrix(~ sex + pmi_c, data = colData(vsd))[ , -1]
)

assay_adj <- cbind(assay_adj, 
                   micro_peak_coords_clean,
                   baseMean = micro_dds_res$baseMean,
                               lfc = micro_dds_res$log2FoldChange,
                               lfcSE = micro_dds_res$lfcSE,
                               #lfc_shr = micro_dds_res_shr$log2FoldChange,
                               stat = micro_dds_res$stat,
                               pvalue = micro_dds_res$pvalue,
                               padj = micro_dds_res$padj,
                   padj_bacon = micro_dds_res_df$padj_bacon)
```

Remove age associated regions from micro pd data
```{r micro_rm_age}
library(GenomeInfoDb)

# convert results ot gr object
micro_dds_gr <-  makeGRangesFromDataFrame(micro_dds_res_df %>%
                                            dplyr::select(-any_of(c("seqnames", "ranges", "strand", "seqlevels", "seqlengths", "isCircular", 
                                                                    "width", "element"))),
                                          keep.extra.columns = TRUE,
                                          seqnames.field = "chr",
                                          start.field = "start",
                                          end.field = "end")

# put both in same style
seqlevelsStyle(micro_dds_gr) <- "UCSC"
seqlevelsStyle(micro_age_sig_gr) <- "UCSC"

# find overlaps with age tt
micro_overlaps <- findOverlaps(micro_dds_gr, 
                               micro_age_sig_gr, 
                               ignore.strand = TRUE) 


# remove overlaps
micro_dds_res_df_noage <- micro_dds_res_df[!rownames(micro_dds_res_df) %in% (micro_overlaps@from), ]

write.csv(micro_dds_res_df_noage,
          file = here("data_out",
                      "080_diff_analysis",
                      "20251112_micro_deseq2_res.csv"))
```

```{r subset_micro_hyper_hypo}
# get only sig hyper acetylated peaks
micro_dds_res_df_hyper <- micro_dds_res_df_noage %>%
  subset(padj < 0.05 & lfc > 0) 

nrow(micro_dds_res_df_hyper)

# get only sig hypoacetylated peaks
micro_dds_res_df_hypo <- micro_dds_res_df_noage %>%
  subset(padj < 0.05 & lfc < 0) 

nrow(micro_dds_res_df_hypo)
```

```{r micro_qqplot}
micro_qqplot <- qq_p_gg(micro_dds_res_df$pvalue_bacon) +
  # ggtitle("Microglia") +
  annotate("text", 
           label = paste0("λ=", round(lambda_from_p(micro_dds_res_df$pvalue_bacon),2)),
           x = 1, 
           y = 10,
           size = 3)  +
  theme(axis.text = element_text(size = 8),
        axis.title = element_text(size = 10)) 

ggsave(micro_qqplot, 
       filename = here("figures",
                       "20251112_micro_qqplot.png"),
       height = 6.5, width = 8, units = "cm", dpi = 600)
```

```{r micro_volcano}
micro_volcano <- ggplot(micro_dds_res_df_noage, 
                        aes(x = lfc, 
                            y = -log(pvalue_bacon),
                            color = ifelse(padj_bacon < 0.05,"FDR<0.05", "FDR≥0.05"))) +
  geom_point(alpha = 0.8,
             size = 1) +
  theme_bw() +
  scale_color_manual(values = c("#e1235c", "#00326e"),
                     name = "Significance") +
  scale_y_continuous(name = "-log(unadjusted p-value)") +
  scale_x_continuous(name = "log(fold-change)") +
  theme(axis.text = element_text(size = 8),
        axis.title = element_text(size = 10),
        legend.title = element_text(size = 8),
        legend.text = element_text(size = 6), 
        legend.key.height = unit(0.25, "cm"),
        legend.key.width = unit(0.25, "cm"),
        legend.position = "bottom") +
  ggrepel::geom_text_repel(data = micro_dds_res_df_noage %>% 
                             arrange(padj_bacon) %>% 
                             slice_head(n=10),
                           aes(label = SYMBOL,),
                           max.overlaps = 100,
                           box.padding = 0.4, 
                           min.segment.length = unit(0, 'lines'),
                           color = "black", 
                           size = 2.5) 

ggsave(micro_volcano, 
       filename = here("figures",
                       "20251112_micro_volcano_legend.png"),
       height = 7, width = 8, units = "cm", dpi = 600)
```

```{r micro_big_wigs}
# import microglia control and pd bigwigs
micro_pd_gr <- rtracklayer::import(here("data_out",
                                        "070_peaks", 
                                        "079_bigwigs_groups",
                                        "microglia",
                                        "microglia_disease_merged_cpm.bw"), 
                                   format = "bigWig")

micro_ctl_gr <- rtracklayer::import(here("data_out",
                                         "070_peaks", 
                                         "079_bigwigs_groups",
                                         "microglia",
                                         "microglia_control_merged_cpm.bw"), 
                                    format = "bigWig")
```

```{r micro_vis_KMT2B_track}
# region (use NCBI-style chr names for hg38)
gen   <- "hg38"
chrom <- 19
start <- 35717103
end   <- 35739640


ax <- GenomeAxisTrack()

mart <- useEnsembl(biomart = "genes", 
                   dataset = "hsapiens_gene_ensembl")

# get gene information for this region
grt <- BiomartGeneRegionTrack(chromosome= chrom,
                              start = start,
                              end = end,
                              genome = gen,
                              name = "Ensembl (HGNC symbols)",
                              biomart = mart)

rg <- ranges(grt)
sym <- mcols(rg)$symbol
gid <- mcols(rg)$gene
sym[is.na(sym) | sym == ""] <- gid[is.na(sym) | sym == ""]
mcols(rg)$symbol <- sym
grt <- GeneRegionTrack(rg, genome = gen, chromosome = chrom, name = "Genes")

# set gene track
displayPars(grt) <- list(
  transcriptAnnotation = "symbol",  # show HGNC symbols
  geneSymbol = TRUE,
  collapseTranscripts = "meta",     # "meta" | "longest" | "gene" | FALSE
  fontsize = 10
)


# get control signal track
control_tr <- DataTrack(
  range       = micro_ctl_gr,
  genome      = gen,
  chromosome  = chrom,
  type        = "histogram",
  col.histogram="#00326e",
  fill.histogram="#00326e",
  name        = "Control (CPM)",
)

# get pd signal track
pd_tr <- DataTrack(
  range       = micro_pd_gr,
  genome      = gen,
  chromosome  = chrom,
  type        = "histogram",
  col.histogram="#e1235c", 
  fill.histogram="#e1235c",
  name        = "PD (CPM)"
)

# create track to indicate where peak
peak_start <- 35727103
peak_end   <- 35729640

# Build a clean GRanges with an 'id' column (character, length == n features)
diff_peak <- GRanges(seqnames = chrom, ranges = IRanges(peak_start, peak_end))
mcols(diff_peak)$id <- NA  # label to display

# Sanity check: id length matches feature count
stopifnot(length(mcols(diff_peak)$id) == length(diff_peak))

diff_tr <- AnnotationTrack(
  range      = diff_peak,
  genome     = gen,
  chromosome = chrom,
  col        = "black",
  fill = "black",
  lwd        = 2,
  stacking   = "dense"
)

displayPars(diff_tr) <- c(
  displayPars(diff_tr),
  list(
    featureAnnotation = NULL,  # draw text from mcols(...)$id
    fontsize          = 9,
    just              = "above"  # place label above feature
  )
)

# get chromosome track
ideo_tr <- IdeogramTrack(genome = gen, chromosome = chrom)

# plot tracks together 
png(here("figures", "micro_kmt2b_track.png"), width = 3600, height = 3600, res = 600)

plotTracks(list(ideo_tr, 
                ax, 
                grt, 
                pd_tr, 
                control_tr, 
                diff_tr),
           chromosome = chrom, 
           from = start,
           to = end,
           background.title = "white",
           col.title = "black",
           cex.axis = 0.6,   # bigger axis tick labels
           cex.title = 0.8,  # track titles
           cex.main = 0.8,
            sizes = c(0.25, 1, 1, 2, 2, 0.25))    # if you use main = "...")

dev.off()
```

```{r micro_vis_KMT2B_boxlot}
micro_kmt2b_boxplot <- micro_dds_res_df_noage %>%
  arrange(padj_bacon) %>%
  slice_head(n=1) %>%
  dplyr::select("IGF136820":"IGF136844") %>%
  pivot_longer(everything(),
               names_to = "sample",
               values_to = "counts") %>%
  merge(., 
        micro_meta,
        by = "sample") %>%
  ggplot(aes(x=group,
             y = counts, 
             fill = group)) +
  geom_boxplot(alpha = 0.8) +
  geom_point() +
  theme_bw() +
  theme(axis.text = element_text(size = 8),
        axis.title.y = element_text(size = 10),
        axis.title.x = element_blank(),
        legend.position = "none") +
  scale_fill_manual(values = c("#00326e", "#e1235c"))

ggsave(plot = micro_kmt2b_boxplot, device = "png",
       filename = here("figures", "20251112_micro_kmt2b_boxplot.png"),
       height = 8, width = 8, units = "cm", dpi = 600)

```


```{r micro_vis_adarb2_1_track}
# region (use NCBI-style chr names for hg38)
gen   <- "hg38"
chrom <- 10
start <- 1647589
end   <- 1669739

ax <- GenomeAxisTrack()

mart <- useEnsembl(biomart = "genes", 
                   dataset = "hsapiens_gene_ensembl")

# get gene information for this region
grt <- BiomartGeneRegionTrack(chromosome= chrom,
                              start = start,
                              end = end,
                              genome = gen,
                              name = "Ensembl (HGNC symbols)",
                              biomart = mart)

rg <- ranges(grt)
sym <- mcols(rg)$symbol
gid <- mcols(rg)$gene
sym[is.na(sym) | sym == ""] <- gid[is.na(sym) | sym == ""]
mcols(rg)$symbol <- sym
grt <- GeneRegionTrack(rg, genome = gen, chromosome = chrom, name = "Genes")

# set gene track
displayPars(grt) <- list(
  transcriptAnnotation = "symbol",  # show HGNC symbols
  geneSymbol = TRUE,
  collapseTranscripts = "meta",     # "meta" | "longest" | "gene" | FALSE
  fontsize = 10
)


# get control signal track
control_tr <- DataTrack(
  range       = micro_ctl_gr,
  genome      = gen,
  chromosome  = chrom,
  type        = "histogram",
  col.histogram="#00326e",
  fill.histogram="#00326e",
  name        = "Control (CPM)",
)

# get pd signal track
pd_tr <- DataTrack(
  range       = micro_pd_gr,
  genome      = gen,
  chromosome  = chrom,
  type        = "histogram",
  col.histogram="#e1235c", 
  fill.histogram="#e1235c",
  name        = "PD (CPM)"
)

# create track to indicate where peak
peak_start <- 1657589
peak_end   <- 1659739

# Build a clean GRanges with an 'id' column (character, length == n features)
diff_peak <- GRanges(seqnames = chrom, ranges = IRanges(peak_start, peak_end))
mcols(diff_peak)$id <- NA  # label to display

# Sanity check: id length matches feature count
stopifnot(length(mcols(diff_peak)$id) == length(diff_peak))

diff_tr <- AnnotationTrack(
  range      = diff_peak,
  genome     = gen,
  chromosome = chrom,
  col        = "black",
  fill = "black",
  lwd        = 2,
  stacking   = "dense"
)

displayPars(diff_tr) <- c(
  displayPars(diff_tr),
  list(
    featureAnnotation = NULL,  # draw text from mcols(...)$id
    fontsize          = 9,
    just              = "above"  # place label above feature
  )
)

# get chromosome track
ideo_tr <- IdeogramTrack(genome = gen, chromosome = chrom)

# plot tracks together 
png(here("figures", "micro_kmt2b_track.png"), width = 3600, height = 3600, res = 600)

plotTracks(list(ideo_tr, 
                ax, 
                grt, 
                pd_tr, 
                control_tr, 
                diff_tr),
           chromosome = chrom, 
           from = start,
           to = end,
           background.title = "white",
           col.title = "black",
           cex.axis = 0.6,   # bigger axis tick labels
           cex.title = 0.8,  # track titles
           cex.main = 0.8)    # if you use main = "...")

dev.off()
```

```{r micro_vis_adarb2_1_boxlot}
micro_dds_res_df_noage %>%
  subset(padj_bacon < 0.05) %>%
  arrange(padj_bacon) %>%
  slice_head(n=2) %>%
   slice_tail(n=1) %>%
  dplyr::select("IGF136820":"IGF136844") %>%
  pivot_longer(everything(),
               names_to = "sample",
               values_to = "counts") %>%
  merge(., 
        micro_meta,
        by = "sample") %>%
  ggplot(aes(x=group,
             y = counts, 
             fill = group)) +
  geom_boxplot(alpha = 0.8) +
  geom_point() +
  theme_bw() +
  theme(axis.text = element_text(size = 8),
        axis.title.y = element_text(size = 10),
        axis.title.x = element_blank(),
        legend.position = "none") +
  scale_fill_manual(values = c("#00326e", "#e1235c"))

```


```{r micro_vis_adarb2_2_track}
# region (use NCBI-style chr names for hg38)
gen   <- "hg38"
chrom <- 10
start <- 1640589
end   <- 1662567

ax <- GenomeAxisTrack()

mart <- useEnsembl(biomart = "genes", 
                   dataset = "hsapiens_gene_ensembl")

# get gene information for this region
grt <- BiomartGeneRegionTrack(chromosome= chrom,
                              start = start,
                              end = end,
                              genome = gen,
                              name = "Ensembl (HGNC symbols)",
                              biomart = mart)

rg <- ranges(grt)
sym <- mcols(rg)$symbol
gid <- mcols(rg)$gene
sym[is.na(sym) | sym == ""] <- gid[is.na(sym) | sym == ""]
mcols(rg)$symbol <- sym
grt <- GeneRegionTrack(rg, genome = gen, chromosome = chrom, name = "Genes")

# set gene track
displayPars(grt) <- list(
  transcriptAnnotation = "symbol",  # show HGNC symbols
  geneSymbol = TRUE,
  collapseTranscripts = "meta",     # "meta" | "longest" | "gene" | FALSE
  fontsize = 10
)


# get control signal track
control_tr <- DataTrack(
  range       = micro_ctl_gr,
  genome      = gen,
  chromosome  = chrom,
  type        = "histogram",
  col.histogram="#00326e",
  fill.histogram="#00326e",
  name        = "Control (CPM)",
)

# get pd signal track
pd_tr <- DataTrack(
  range       = micro_pd_gr,
  genome      = gen,
  chromosome  = chrom,
  type        = "histogram",
  col.histogram="#e1235c", 
  fill.histogram="#e1235c",
  name        = "PD (CPM)"
)

# create track to indicate where peak
peak_start <- 1650589
peak_end   <- 1652567

	

# Build a clean GRanges with an 'id' column (character, length == n features)
diff_peak <- GRanges(seqnames = chrom, ranges = IRanges(peak_start, peak_end))
mcols(diff_peak)$id <- NA  # label to display

# Sanity check: id length matches feature count
stopifnot(length(mcols(diff_peak)$id) == length(diff_peak))

diff_tr <- AnnotationTrack(
  range      = diff_peak,
  genome     = gen,
  chromosome = chrom,
  col        = "black",
  fill = "black",
  lwd        = 2,
  stacking   = "dense"
)

displayPars(diff_tr) <- c(
  displayPars(diff_tr),
  list(
    featureAnnotation = NULL,  # draw text from mcols(...)$id
    fontsize          = 9,
    just              = "above"  # place label above feature
  )
)

# get chromosome track
ideo_tr <- IdeogramTrack(genome = gen, chromosome = chrom)

# plot tracks together 
png(here("figures", "micro_kmt2b_track.png"), width = 3600, height = 3600, res = 600)

plotTracks(list(ideo_tr, 
                ax, 
                grt, 
                pd_tr, 
                control_tr, 
                diff_tr),
           chromosome = chrom, 
           from = start,
           to = end,
           background.title = "white",
           col.title = "black",
           cex.axis = 0.6,   # bigger axis tick labels
           cex.title = 0.8,  # track titles
           cex.main = 0.8)    # if you use main = "...")

dev.off()
```

```{r micro_vis_adarb2_2_boxlot}
micro_dds_res_df_noage %>%
  subset(padj_bacon < 0.05) %>%
  arrange(padj_bacon) %>%
  slice_head(n=3) %>%
   slice_tail(n=1) %>%
  dplyr::select("IGF136820":"IGF136844") %>%
  pivot_longer(everything(),
               names_to = "sample",
               values_to = "counts") %>%
  merge(., 
        micro_meta,
        by = "sample") %>%
  ggplot(aes(x=group,
             y = counts, 
             fill = group)) +
  geom_boxplot(alpha = 0.8) +
  geom_point() +
  theme_bw() +
  theme(axis.text = element_text(size = 8),
        axis.title.y = element_text(size = 10),
        axis.title.x = element_blank(),
        legend.position = "none") +
  scale_fill_manual(values = c("#00326e", "#e1235c"))

```

```{r micro_recap_rat_rotenone}
# bcl 
micro_bcl <- micro_dds_res_df_noage %>% 
  subset(SYMBOL == "BCL3") %>%
  subset(grepl("Promoter", annotation)) %>%
  dplyr::select(c(chr, start, end, IGF136820:IGF136844)) %>%
  pivot_longer(-c(chr,start,end), names_to = "sample", values_to = "counts") %>% 
  merge(., micro_meta, by = "sample") %>% 
  mutate(name = paste0("chr", chr, ":", start, "-", end)) %>%
  ggplot(aes(x = group, y = counts, fill = group)) +
  geom_boxplot(alpha = 0.8) +
  geom_point() +
  facet_wrap(~name, scales = "free_y", nrow = 1) +
  theme_bw() +
  theme(axis.text = element_text(size = 8),
        axis.title.y = element_text(size = 10),
        axis.title.x = element_blank(),
        legend.position = "none", 
        strip.text = element_text(size = 6)) +
  scale_fill_manual(values = c("#00326e", "#e1235c")) +
  scale_y_continuous(name = "Counts") + 
  scale_x_discrete(labels = c("Control", "PD"))

ggsave(micro_bcl, filename = here("figures", "20251112_micro_bcl3.png"),
       height = 5, width = 20, units = "cm", dpi = 600)

# mat2b 
micro_mat2b <- micro_dds_res_df_noage %>% 
  subset(SYMBOL == "MAT2B") %>%
  subset(grepl("Distal Intergenic", annotation)) %>%
  dplyr::select(c(chr, start, end, IGF136820:IGF136844)) %>%
  pivot_longer(-c(chr,start,end), names_to = "sample", values_to = "counts") %>% 
  merge(., micro_meta, by = "sample") %>% 
  mutate(name = paste0("chr", chr, ":", start, "-", end)) %>%
  ggplot(aes(x = group, y = counts, fill = group)) +
  geom_boxplot(alpha = 0.8) +
  geom_point() +
  facet_wrap(~name, scales = "free_y") +
  theme_bw() +
  theme(axis.text = element_text(size = 8),
        axis.title.y = element_text(size = 10),
        axis.title.x = element_blank(),
        legend.position = "none", 
        strip.text = element_text(size = 6)) +
  scale_fill_manual(values = c("#00326e", "#e1235c")) +
  scale_y_continuous(name = "Counts") + 
  scale_x_discrete(labels = c("Control", "PD"))

ggsave(micro_mat2b, filename = here("figures", "20251112_micro_mat2b.png"),
       height = 5, width = 5, units = "cm", dpi = 600)

# hif3a 
micro_hif3a <- micro_dds_res_df_noage %>% 
  subset(SYMBOL == "HIF3A") %>%
  subset(grepl("Promoter", annotation)) %>%
  dplyr::select(c(chr, start, end, IGF136820:IGF136844)) %>%
  pivot_longer(-c(chr,start,end), names_to = "sample", values_to = "counts") %>% 
  merge(., micro_meta, by = "sample") %>% 
  mutate(name = paste0("chr", chr, ":", start, "-", end)) %>%
  ggplot(aes(x = group, y = counts, fill = group)) +
  geom_boxplot(alpha = 0.8) +
  geom_point() +
  facet_wrap(~name, scales = "free_y") +
  theme_bw() +
  theme(axis.text = element_text(size = 8),
        axis.title.y = element_text(size = 10),
        axis.title.x = element_blank(),
        legend.position = "none", 
        strip.text = element_text(size = 6)) +
  scale_fill_manual(values = c("#00326e", "#e1235c")) +
  scale_y_continuous(name = "Counts") + 
  scale_x_discrete(labels = c("Control", "PD"))

ggsave(micro_hif3a, filename = here("figures", "20251112_micro_hif3a.png"),
       height = 5, width = 12, units = "cm", dpi = 600)

# olfm4 
micro_olfm4 <- micro_dds_res_df_noage %>% 
  subset(SYMBOL == "OLFM4") %>%
  #subset(grepl("Promoter", annotation)) %>%
  dplyr::select(c(chr, start, end, IGF136820:IGF136844)) %>%
  pivot_longer(-c(chr,start,end), names_to = "sample", values_to = "counts") %>% 
  merge(., micro_meta, by = "sample") %>% 
  mutate(name = paste0("chr", chr, ":", start, "-", end)) %>%
  ggplot(aes(x = group, y = counts, fill = group)) +
  geom_boxplot(alpha = 0.8) +
  geom_point() +
  facet_wrap(~name, scales = "free_y", nrow = 1) +
  theme_bw() +
  theme(axis.text = element_text(size = 8),
        axis.title.y = element_text(size = 10),
        axis.title.x = element_blank(),
        legend.position = "none", 
        strip.text = element_text(size = 6)) +
  scale_fill_manual(values = c("#00326e", "#e1235c")) +
  scale_y_continuous(name = "Counts") + 
  scale_x_discrete(labels = c("Control", "PD"))

ggsave(micro_olfm4, filename = here("figures", "20251112_micro_olfm4.png"),
       height = 5, width = 25, units = "cm", dpi = 600)
```

```{r vis_top_peak_hypo}
# region (use NCBI-style chr names for hg38)
gen   <- "hg38"
chrom <- 11
start <- 31792734
end   <- 31819119

ax <- GenomeAxisTrack()

mart <- useEnsembl(biomart = "genes", 
                   dataset = "hsapiens_gene_ensembl")

grt <- BiomartGeneRegionTrack(
  chromosome= chrom,
  start = start,
  end = end,
  genome = gen,
  name        = "Ensembl (HGNC symbols)",
  biomart     = mart
)

rg <- ranges(grt)
sym <- mcols(rg)$symbol
gid <- mcols(rg)$gene
sym[is.na(sym) | sym == ""] <- gid[is.na(sym) | sym == ""]
mcols(rg)$symbol <- sym
grt <- GeneRegionTrack(rg, genome = gen, chromosome = chrom, name = "Genes")


displayPars(grt) <- list(
  transcriptAnnotation = "symbol",  # show HGNC symbols
  geneSymbol = TRUE,
  collapseTranscripts = "meta",     
  fontsize = 10
)

# signal trakcs
control_tr <- DataTrack(
  range       = ctl_bw,
  genome      = gen,
  chromosome  = chrom,
  type        = "histogram",
  col.histogram="darkblue",
  fill.histogram="darkblue",
  name        = "Control (CPM)",
)

pd_tr <- DataTrack(
  range       = dis_bw,
  genome      = gen,
  chromosome  = chrom,
  type        = "histogram",
  col.histogram="darkred", 
  fill.histogram="darkred",
  name        = "Disease (CPM)"
)


# define the differential peak span 
peak_start <- 31802734
peak_end   <- 31809119

# Build a clean GRanges with an 'id' column 
diff_peak <- GRanges(seqnames = chrom, ranges = IRanges(peak_start, peak_end))
mcols(diff_peak)$id <- NA  # label to display

# Sanity check: id length matches feature count
stopifnot(length(mcols(diff_peak)$id) == length(diff_peak))

# Make the boxed AnnotationTrack
diff_tr <- AnnotationTrack(
  range      = diff_peak,
  genome     = gen,
  chromosome = chrom,
  col        = "black",
  fill = "black",
  lwd        = 2,
  stacking   = "dense"
)


displayPars(diff_tr) <- c(
  displayPars(diff_tr),
  list(
    featureAnnotation = NULL,  # draw text from mcols(...)$id
    fontsize          = 9,
    just              = "above"  # place label above feature
  )
)

# plot tracks t
plotTracks(list(ax, grt, pd_tr, control_tr, diff_tr),
           chromosome = chrom, from = start, to = end,
           background.title = "white", col.title = "black")
```


```{r micro_go_terms_hyper}
# convert peaks to granges obj
micro_dds_res_df_hyper_gr <- GRanges(seqnames = micro_dds_res_df_hyper$chr,
                                     ranges   = IRanges(as.numeric(micro_dds_res_df_hyper$start), 
                                                        as.numeric(micro_dds_res_df_hyper$end)),
                                     strand   = "*")

# harmonize chromosome naming style with TxDb (e.g., "chr1" vs "1")
seqlevelsStyle(micro_dds_res_df_hyper_gr) <- seqlevelsStyle(TxDb.Hsapiens.UCSC.hg38.knownGene)

# annotate peaks
micro_dds_res_df_hyper_anno <- annotatePeak(micro_dds_res_df_hyper_gr, 
                                            TxDb=TxDb.Hsapiens.UCSC.hg38.knownGene, 
                                            tssRegion=c(-3000,3000), 
                                            annoDb="org.Hs.eg.db")

# merge annotate to res df
micro_dds_res_df_hyper <- merge(micro_dds_res_df_hyper,
                                as.data.frame(micro_dds_res_df_hyper_anno@anno) %>%
                                  mutate(chr = str_sub(seqnames, 4, 6)),
                                by = c("chr", "start", "end"))

micro_hyper_genes <- unique(na.omit(as.data.frame(micro_dds_res_df_hyper_anno)$ENSEMBL))

micro_hyper_go <- enrichGO(gene=micro_hyper_genes, 
                           OrgDb=org.Hs.eg.db, 
                           keyType="ENSEMBL",
                           ont="BP", 
                           #pvalueCutoff=0.05, 
                           qvalueCutoff = 0.5)
dotplot(micro_hyper_go)
```

```{r micro_go_terms_hypo}
# convert peaks to granges obj
micro_dds_res_df_hypo_gr <- GRanges(seqnames = micro_dds_res_df_hypo$chr,
                                    ranges   = IRanges(as.numeric(micro_dds_res_df_hypo$start), 
                                                       as.numeric(micro_dds_res_df_hypo$end)),
                                    strand   = "*")

# harmonize chromosome naming style with TxDb (e.g., "chr1" vs "1")
seqlevelsStyle(micro_dds_res_df_hypo_gr) <- seqlevelsStyle(TxDb.Hsapiens.UCSC.hg38.knownGene)

# annotate peaks
micro_dds_res_df_hypo_anno <- annotatePeak(micro_dds_res_df_hypo_gr, 
                                           TxDb=TxDb.Hsapiens.UCSC.hg38.knownGene, 
                                           tssRegion=c(-3000,3000), 
                                           annoDb="org.Hs.eg.db")

# merge annotate to res df
micro_dds_res_df_hypo <- merge(micro_dds_res_df_hypo,
                               as.data.frame(micro_dds_res_df_hypo_anno@anno) %>%
                                 mutate(chr = str_sub(seqnames, 4, 6)),
                               by = c("chr", "start", "end"))

micro_hypo_genes <- unique(na.omit(as.data.frame(micro_dds_res_df_hypo_anno)$ENSEMBL))

micro_hypo_go <- enrichGO(gene=micro_hypo_genes, 
                          OrgDb=org.Hs.eg.db, 
                          keyType="ENSEMBL",
                          ont="BP", 
                          #pvalueCutoff=0.01, 
                          qvalueCutoff=0.05)
dotplot(micro_hypo_go)
```



### Differential analysis of oligodendrocytes

```{r oligo_diff_filter}
# remove sex chromosomes
oligo_feat_clean_noxy <- oligo_feat_clean_anno %>%
  subset(!chr %in% c("X", "Y"))

# separate coordinates and counts 
oligo_peak_coords <- oligo_feat_clean_noxy[, c(1:3, 13:27)]
oligo_counts <- as.matrix(oligo_feat_clean_noxy[, -c(1:3, 13:27)])

# recode group as factor
# set control as base level
oligo_meta <- oligo_meta %>%
  mutate(group = relevel(factor(group), 
                         ref = "control")) 

# create edgeR dgelist for filtering
oligo_dgelist <- DGEList(counts = oligo_counts, 
                         samples = oligo_meta)

# replicate-aware filtering - min count 5 
oligo_keep <- filterByExpr(oligo_dgelist, 
                           group = oligo_dgelist$samples$group, 
                           min.count = 5, 
                           min.total.count=50) # smallest group (control) has 4 samples

sum(oligo_keep) #

# remove those with <5 from count and peaks
oligo_peak_coords_clean <- oligo_peak_coords[oligo_keep, ]
oligo_counts_clean <- oligo_counts[oligo_keep, ]

# make deseq2 object
oligo_dds = DESeqDataSetFromMatrix(countData = oligo_counts_clean,
                                   colData = oligo_meta %>%
                                     mutate(pmi_c = scale(pmi, center = T, scale = F)),
                                   design = ~ group + pmi + sex)
```

```{r oligo_check_vif}
md <- as.data.frame(colData(oligo_dds))

# Ensure factors are coded as factors; pmi numeric
md$pd  <- as.factor(md$group)
md$sex <- as.factor(md$sex)
md$pmi <- as.numeric(md$pmi)

# Fit a dummy linear model (response is random noise)
set.seed(1)
fit <- lm(rnorm(nrow(md)) ~ group + pmi + sex, data = md)

# VIFs (GVIF for multi-df factors)
vif(fit)
#  group      pmi      sex 
# 1.036150 1.607185 1.597796 

```

```{r oligo_diff_deseq2}
# diff analysis using deseq2
oligo_dds_diff = DESeq(oligo_dds, 
                       test = "Wald", 
                       fitType = "local",
                       # reduced = ~ sv1, # lrt only
                       sfType = "poscounts")

# get normalized peak counts (uses method from DESeq)
oligo_dds_norm_counts = counts(oligo_dds_diff, 
                               normalized = TRUE)  

# create results dataframe
oligo_dds_res = results(oligo_dds_diff, 
                        pAdjustMethod = "BH",
                        independentFiltering = FALSE, 
                        altHypothesis = "greaterAbs")

# get result names
# resultsNames(oligo_dds_diff)

# # fold change shrinkage
# oligo_dds_res_shr <- lfcShrink(oligo_dds_diff, 
#                                coef = "group_pd_vs_control", 
#                                type = "apeglm")

# make results df
oligo_dds_res_df <- data.frame(
  baseMean = oligo_dds_res$baseMean,
  lfc = oligo_dds_res$log2FoldChange,
  lfcSE = oligo_dds_res$lfcSE,
  # lfc_shr = oligo_dds_res_shr$log2FoldChange,
  stat = oligo_dds_res$stat,
  pvalue = oligo_dds_res$pvalue,
  padj = oligo_dds_res$padj,
  oligo_peak_coords_clean, 
  oligo_dds_norm_counts) 

bc <- bacon(oligo_dds_res_df$stat)
oligo_dds_res_df$padj_bacon <- pval(bc) %>% p.adjust(method = "BH")
oligo_dds_res_df$pvalue_bacon <- pval(bc)
```

Remove age associated regions from oligo pd data
```{r oligo_rm_age}
# convert results ot gr object
oligo_dds_gr <-  makeGRangesFromDataFrame(oligo_dds_res_df %>%
                                            dplyr::select(-any_of(c("seqnames", "ranges", "strand", "seqlevels", "seqlengths", "isCircular", 
                                                                    "width", "element"))),
                                          keep.extra.columns = TRUE,
                                          seqnames.field = "chr",
                                          start.field = "start",
                                          end.field = "end")

# put both in same style
seqlevelsStyle(oligo_dds_gr) <- "UCSC"
seqlevelsStyle(oligo_age_sig_gr) <- "UCSC"

# find overlaps with age tt
oligo_overlaps <- findOverlaps(oligo_dds_gr, 
                               oligo_age_sig_gr, 
                               ignore.strand = TRUE) 


# remove overlaps
oligo_dds_res_df_noage <- oligo_dds_res_df[!rownames(oligo_dds_res_df) %in% (oligo_overlaps@from), ]

write.csv(oligo_dds_res_df_noage,
          file = here("data_out",
                      "080_diff_analysis",
                      "20251112_oligo_deseq2_res.csv"))
```

```{r subset_oligo_hyper_hypo}
# get only sig hyper acetylated peaks
oligo_dds_res_df_hyper <- oligo_dds_res_df_noage %>%
  subset(padj < 0.05 & lfc > 0) 

nrow(oligo_dds_res_df_hyper)

# get only sig hypoacetylated peaks
oligo_dds_res_df_hypo <- oligo_dds_res_df_noage %>%
  subset(padj < 0.05 & lfc < 0) 

nrow(oligo_dds_res_df_hypo)
```

```{r oligo_qqplot}
oligo_qqplot <- qq_p_gg(oligo_dds_res_df$pvalue_bacon) +
  # ggtitle("oligoglia") +
  annotate("text", 
           label = paste0("λ=", round(lambda_from_p(oligo_dds_res_df$pvalue_bacon),2)),
           x = 4, 
           y = 2,
           size = 3)  +
  theme(axis.text = element_text(size = 8),
        axis.title = element_text(size = 10)) 

ggsave(oligo_qqplot, 
       filename = here("figures",
                       "20251112_oligo_qqplot.png"),
       height = 6.5, width = 8, units = "cm", dpi = 600)
```

```{r oligo_volcano}
oligo_volcano <- ggplot(oligo_dds_res_df, 
                        aes(x = lfc, 
                            y = -log(pvalue_bacon),
                            color = ifelse(padj_bacon < 0.05,"FDR<0.05", "FDR≥0.05"))) +
  geom_point(alpha = 0.8,
             size = 1) +
  theme_bw() +
  scale_color_manual(values = c("#e1235c", "#00326e"),
                     name = "Significance") +
  scale_y_continuous(name = "-log(unadjusted p-value)") +
  scale_x_continuous(name = "log(fold-change)") +
  theme(axis.text = element_text(size = 8),
        axis.title = element_text(size = 10),
        legend.title = element_text(size = 9),
        legend.text = element_text(size = 8), 
        legend.key.height = unit(0.25, "cm"),
        legend.key.width = unit(0.25, "cm"),
        legend.position = "bottom") +
  ggrepel::geom_text_repel(data = oligo_dds_res_df %>% 
                             arrange(padj_bacon) %>% 
                             slice_head(n=10),
                           aes(label = SYMBOL),
                           max.overlaps = 100,
                           box.padding = 0.4, 
                           min.segment.length = unit(0, 'lines'),
                           color = "black", 
                           size = 2.5) 

ggsave(oligo_volcano, 
       filename = here("figures",
                       "20251112_oligo_volcano_legend.png"),
       height = 7, width = 8, units = "cm", dpi = 600)
```

```{r oligo_big_wigs}
# import oligoglia control and pd bigwigs
oligo_pd_gr <- rtracklayer::import(here("data_out",
                                        "070_peaks", 
                                        "079_bigwigs_groups",
                                        "oligodendrocyte",
                                        "oligodendrocyte_disease_merged_cpm.bw"), 
                                   format = "bigWig")

oligo_ctl_gr <- rtracklayer::import(here("data_out",
                                         "070_peaks", 
                                         "079_bigwigs_groups",
                                         "oligodendrocyte",
                                         "oligodendrocyte_control_merged_cpm.bw"), 
                                    format = "bigWig")
```

```{r oligo_vis_KMT2B_track}
# region (use NCBI-style chr names for hg38)
gen   <- "hg38"
chrom <- 19
start <- 35717008
end   <- 35740790

ax <- GenomeAxisTrack()

mart <- useEnsembl(biomart = "genes", 
                   dataset = "hsapiens_gene_ensembl")

# get gene information for this region
grt <- BiomartGeneRegionTrack(chromosome= chrom,
                              start = start,
                              end = end,
                              genome = gen,
                              name = "Ensembl (HGNC symbols)",
                              biomart = mart)

rg <- ranges(grt)
sym <- mcols(rg)$symbol
gid <- mcols(rg)$gene
sym[is.na(sym) | sym == ""] <- gid[is.na(sym) | sym == ""]
mcols(rg)$symbol <- sym
grt <- GeneRegionTrack(rg, genome = gen, chromosome = chrom, name = "Genes")

# set gene track
displayPars(grt) <- list(
  transcriptAnnotation = "symbol",  # show HGNC symbols
  geneSymbol = TRUE,
  collapseTranscripts = "meta",     # "meta" | "longest" | "gene" | FALSE
  fontsize = 10
)


# get control signal track
control_tr <- DataTrack(
  range       = oligo_ctl_gr,
  genome      = gen,
  chromosome  = chrom,
  type        = "histogram",
  col.histogram="#00326e",
  fill.histogram="#00326e",
  name        = "Control (CPM)",
)

# get pd signal track
pd_tr <- DataTrack(
  range       = oligo_pd_gr,
  genome      = gen,
  chromosome  = chrom,
  type        = "histogram",
  col.histogram="#e1235c", 
  fill.histogram="#e1235c",
  name        = "PD (CPM)"
)

# create track to indicate where peak
peak_start <- 35727008
peak_end   <- 35730790

# Build a clean GRanges with an 'id' column (character, length == n features)
diff_peak <- GRanges(seqnames = chrom, ranges = IRanges(peak_start, peak_end))
mcols(diff_peak)$id <- NA  # label to display

# Sanity check: id length matches feature count
stopifnot(length(mcols(diff_peak)$id) == length(diff_peak))

diff_tr <- AnnotationTrack(
  range      = diff_peak,
  genome     = gen,
  chromosome = chrom,
  col        = "black",
  fill = "black",
  lwd        = 2,
  stacking   = "dense"
)

displayPars(diff_tr) <- c(
  displayPars(diff_tr),
  list(
    featureAnnotation = NULL,  # draw text from mcols(...)$id
    fontsize          = 9,
    just              = "above"  # place label above feature
  )
)

# get chromosome track
ideo_tr <- IdeogramTrack(genome = gen, chromosome = chrom)

# plot tracks together 
png(here("figures", "oligo_kmt2b_track.png"), width = 3600, height = 3600, res = 600)

plotTracks(list(ideo_tr, 
                ax, 
                grt, 
                pd_tr, 
                control_tr, 
                diff_tr),
           chromosome = chrom, 
           from = start,
           to = end,
           background.title = "white",
           col.title = "black",
           cex.axis = 0.6,   # bigger axis tick labels
           cex.title = 0.8,  # track titles
           cex.main = 0.8,
            sizes = c(0.25, 1, 1, 2, 2, 0.25))    # if you use main = "...")

dev.off()
```

```{r oligo_vis_KMT2B_boxlot}
oligo_dds_res_df_noage %>%
  subset(padj_bacon < 0.05) %>%
  arrange(padj_bacon) %>%
  slice_head(n=1) %>%
  dplyr::select("IGF136819":"IGF138590") %>%
  pivot_longer(everything(),
               names_to = "sample",
               values_to = "counts") %>%
  merge(., 
        oligo_meta,
        by = "sample") %>%
  ggplot(aes(x=group,
             y = counts, 
             fill = group)) +
  geom_boxplot(alpha = 0.8) +
  geom_point() +
  theme_bw() +
  theme(axis.text = element_text(size = 8),
        axis.title.y = element_text(size = 10),
        axis.title.x = element_blank(),
        legend.position = "none") +
  scale_fill_manual(values = c("#00326e", "#e1235c"))
```



```{r oligo_vis_MIR1257_track}
# region (use NCBI-style chr names for hg38)
gen   <- "hg38"
chrom <- 20
start <- 61863651
end   <- 61886844

ax <- GenomeAxisTrack()

mart <- useEnsembl(biomart = "genes", 
                   dataset = "hsapiens_gene_ensembl")

# get gene information for this region
grt <- BiomartGeneRegionTrack(chromosome= chrom,
                              start = start,
                              end = end,
                              genome = gen,
                              name = "Ensembl (HGNC symbols)",
                              biomart = mart)

rg <- ranges(grt)
sym <- mcols(rg)$symbol
gid <- mcols(rg)$gene
sym[is.na(sym) | sym == ""] <- gid[is.na(sym) | sym == ""]
mcols(rg)$symbol <- sym
grt <- GeneRegionTrack(rg, genome = gen, chromosome = chrom, name = "Genes")

# set gene track
displayPars(grt) <- list(
  transcriptAnnotation = "symbol",  # show HGNC symbols
  geneSymbol = TRUE,
  collapseTranscripts = "meta",     # "meta" | "longest" | "gene" | FALSE
  fontsize = 10
)


# get control signal track
control_tr <- DataTrack(
  range       = oligo_ctl_gr,
  genome      = gen,
  chromosome  = chrom,
  type        = "histogram",
  col.histogram="#00326e",
  fill.histogram="#00326e",
  name        = "Control (CPM)",
)

# get pd signal track
pd_tr <- DataTrack(
  range       = oligo_pd_gr,
  genome      = gen,
  chromosome  = chrom,
  type        = "histogram",
  col.histogram="#e1235c", 
  fill.histogram="#e1235c",
  name        = "PD (CPM)"
)

# create track to indicate where peak
peak_start <- 61873651
peak_end   <- 61876844

# Build a clean GRanges with an 'id' column (character, length == n features)
diff_peak <- GRanges(seqnames = chrom, ranges = IRanges(peak_start, peak_end))
mcols(diff_peak)$id <- NA  # label to display

# Sanity check: id length matches feature count
stopifnot(length(mcols(diff_peak)$id) == length(diff_peak))

diff_tr <- AnnotationTrack(
  range      = diff_peak,
  genome     = gen,
  chromosome = chrom,
  col        = "black",
  fill = "black",
  lwd        = 2,
  stacking   = "dense"
)

displayPars(diff_tr) <- c(
  displayPars(diff_tr),
  list(
    featureAnnotation = NULL,  # draw text from mcols(...)$id
    fontsize          = 9,
    just              = "above"  # place label above feature
  )
)

# get chromosome track
ideo_tr <- IdeogramTrack(genome = gen, chromosome = chrom)

# plot tracks together 
png(here("figures", "oligo_mir1257_track.png"), width = 3600, height = 3600, res = 600)

plotTracks(list(ideo_tr, 
                ax, 
                grt, 
                pd_tr, 
                control_tr, 
                diff_tr),
           chromosome = chrom, 
           from = start,
           to = end,
           background.title = "white",
           col.title = "black",
           cex.axis = 0.6,   # bigger axis tick labels
           cex.title = 0.8,  # track titles
           cex.main = 0.8,
            sizes = c(0.25, 1, 1, 2, 2, 0.25))    # if you use main = "...")

dev.off()
```

```{r oligo_vis_MIR1257_boxlot}
oligo_dds_res_df_noage %>%
  subset(padj_bacon < 0.05) %>%
  arrange(padj_bacon) %>%
  slice_head(n=2) %>%
  slice_tail(n=1) %>%
  dplyr::select("IGF136819":"IGF138590") %>%
  pivot_longer(everything(),
               names_to = "sample",
               values_to = "counts") %>%
  merge(., 
        oligo_meta,
        by = "sample") %>%
  ggplot(aes(x=group,
             y = counts, 
             fill = group)) +
  geom_boxplot(alpha = 0.8) +
  geom_point() +
  theme_bw() +
  theme(axis.text = element_text(size = 8),
        axis.title.y = element_text(size = 10),
        axis.title.x = element_blank(),
        legend.position = "none") +
  scale_fill_manual(values = c("#00326e", "#e1235c"))
```

```{r oligo_vis_ADARB2_track}
# region (use NCBI-style chr names for hg38)
gen   <- "hg38"
chrom <- 10
start <- 1640536
end   <- 1665314

ax <- GenomeAxisTrack()

mart <- useEnsembl(biomart = "genes", 
                   dataset = "hsapiens_gene_ensembl")

# get gene information for this region
grt <- BiomartGeneRegionTrack(chromosome= chrom,
                              start = start,
                              end = end,
                              genome = gen,
                              name = "Ensembl (HGNC symbols)",
                              biomart = mart)

rg <- ranges(grt)
sym <- mcols(rg)$symbol
gid <- mcols(rg)$gene
sym[is.na(sym) | sym == ""] <- gid[is.na(sym) | sym == ""]
mcols(rg)$symbol <- sym
grt <- GeneRegionTrack(rg, genome = gen, chromosome = chrom, name = "Genes")

# set gene track
displayPars(grt) <- list(
  transcriptAnnotation = "symbol",  # show HGNC symbols
  geneSymbol = TRUE,
  collapseTranscripts = "meta",     # "meta" | "longest" | "gene" | FALSE
  fontsize = 10
)


# get control signal track
control_tr <- DataTrack(
  range       = oligo_ctl_gr,
  genome      = gen,
  chromosome  = chrom,
  type        = "histogram",
  col.histogram="#00326e",
  fill.histogram="#00326e",
  name        = "Control (CPM)",
)

# get pd signal track
pd_tr <- DataTrack(
  range       = oligo_pd_gr,
  genome      = gen,
  chromosome  = chrom,
  type        = "histogram",
  col.histogram="#e1235c", 
  fill.histogram="#e1235c",
  name        = "PD (CPM)"
)

# create track to indicate where peak
peak_start <- 1650536
peak_end   <- 1655314

# Build a clean GRanges with an 'id' column (character, length == n features)
diff_peak <- GRanges(seqnames = chrom, ranges = IRanges(peak_start, peak_end))
mcols(diff_peak)$id <- NA  # label to display

# Sanity check: id length matches feature count
stopifnot(length(mcols(diff_peak)$id) == length(diff_peak))

diff_tr <- AnnotationTrack(
  range      = diff_peak,
  genome     = gen,
  chromosome = chrom,
  col        = "black",
  fill = "black",
  lwd        = 2,
  stacking   = "dense"
)

displayPars(diff_tr) <- c(
  displayPars(diff_tr),
  list(
    featureAnnotation = NULL,  # draw text from mcols(...)$id
    fontsize          = 9,
    just              = "above"  # place label above feature
  )
)

# get chromosome track
ideo_tr <- IdeogramTrack(genome = gen, chromosome = chrom)

# plot tracks together 
png(here("figures", "oligo_adarb2_track.png"), width = 3600, height = 3600, res = 600)

plotTracks(list(ideo_tr, 
                ax, 
                grt, 
                pd_tr, 
                control_tr, 
                diff_tr),
           chromosome = chrom, 
           from = start,
           to = end,
           background.title = "white",
           col.title = "black",
           cex.axis = 0.6,   # bigger axis tick labels
           cex.title = 0.8,  # track titles
           cex.main = 0.8,
            sizes = c(0.25, 1, 1, 2, 2, 0.25))    # if you use main = "...")

dev.off()
```

```{r oligo_vis_ADARB2_boxlot}
oligo_dds_res_df_noage %>%
  subset(padj_bacon < 0.05) %>%
  arrange(padj_bacon) %>%
  slice_head(n=3) %>%
  slice_tail(n=1) %>%
  dplyr::select("IGF136819":"IGF138590") %>%
  pivot_longer(everything(),
               names_to = "sample",
               values_to = "counts") %>%
  merge(., 
        oligo_meta,
        by = "sample") %>%
  ggplot(aes(x=group,
             y = counts, 
             fill = group)) +
  geom_boxplot(alpha = 0.8) +
  geom_point() +
  theme_bw() +
  theme(axis.text = element_text(size = 8),
        axis.title.y = element_text(size = 10),
        axis.title.x = element_blank(),
        legend.position = "none") +
  scale_fill_manual(values = c("#00326e", "#e1235c"))
```



```{r oligo_recap_rat_rotenone}
# bcl 
oligo_bcl <- oligo_dds_res_df %>% 
  subset(SYMBOL == "BCL3") %>%
  subset(grepl("Promoter", annotation)) %>%
  dplyr::select(c(chr, start, end, IGF136819:IGF138590)) %>%
  pivot_longer(-c(chr,start,end), names_to = "sample", values_to = "counts") %>% 
  merge(., oligo_meta, by = "sample") %>% 
  mutate(name = paste0("chr", chr, ":", start, "-", end)) %>%
  ggplot(aes(x = group, y = counts, fill = group)) +
  geom_boxplot(alpha = 0.8) +
  geom_point() +
  facet_wrap(~name, scales = "free_y", nrow = 1) +
  theme_bw() +
  theme(axis.text = element_text(size = 8),
        axis.title.y = element_text(size = 10),
        axis.title.x = element_blank(),
        legend.position = "none", 
        strip.text = element_text(size = 6)) +
  scale_fill_manual(values = c("#00326e", "#e1235c")) +
  scale_y_continuous(name = "Counts") + 
  scale_x_discrete(labels = c("Control", "PD"))

ggsave(oligo_bcl, filename = here("figures", "20251014_oligo_bcl3.png"),
       height = 5, width = 15, units = "cm", dpi = 600)

# mat2b 
oligo_mat2b <- oligo_dds_res_df %>% 
  subset(SYMBOL == "MAT2B") %>%
  subset(grepl("Distal Intergenic", annotation)) %>%
  dplyr::select(c(chr, start, end, IGF136819:IGF138590)) %>%
  pivot_longer(-c(chr,start,end), names_to = "sample", values_to = "counts") %>% 
  merge(., oligo_meta, by = "sample") %>% 
  mutate(name = paste0("chr", chr, ":", start, "-", end)) %>%
  ggplot(aes(x = group, y = counts, fill = group)) +
  geom_boxplot(alpha = 0.8) +
  geom_point() +
  facet_wrap(~name, scales = "free_y") +
  theme_bw() +
  theme(axis.text = element_text(size = 8),
        axis.title.y = element_text(size = 10),
        axis.title.x = element_blank(),
        legend.position = "none", 
        strip.text = element_text(size = 6)) +
  scale_fill_manual(values = c("#00326e", "#e1235c")) +
  scale_y_continuous(name = "Counts") + 
  scale_x_discrete(labels = c("Control", "PD"))

ggsave(oligo_mat2b, filename = here("figures", "20251014_oligo_mat2b.png"),
       height = 5, width = 5, units = "cm", dpi = 600)

# hif3a 
oligo_hif3a <- oligo_dds_res_df %>% 
  subset(SYMBOL == "HIF3A") %>%
  subset(grepl("Promoter", annotation)) %>%
  dplyr::select(c(chr, start, end, IGF136819:IGF138590)) %>%
  pivot_longer(-c(chr,start,end), names_to = "sample", values_to = "counts") %>% 
  merge(., oligo_meta, by = "sample") %>% 
  mutate(name = paste0("chr", chr, ":", start, "-", end)) %>%
  ggplot(aes(x = group, y = counts, fill = group)) +
  geom_boxplot(alpha = 0.8) +
  geom_point() +
  facet_wrap(~name, scales = "free_y") +
  theme_bw() +
  theme(axis.text = element_text(size = 8),
        axis.title.y = element_text(size = 10),
        axis.title.x = element_blank(),
        legend.position = "none", 
        strip.text = element_text(size = 6)) +
  scale_fill_manual(values = c("#00326e", "#e1235c")) +
  scale_y_continuous(name = "Counts") + 
  scale_x_discrete(labels = c("Control", "PD"))

ggsave(oligo_hif3a, filename = here("figures", "20251014_oligo_hif3a.png"),
       height = 5, width = 15, units = "cm", dpi = 600)

# olfm4 
oligo_olfm4 <- oligo_dds_res_df %>% 
  subset(SYMBOL == "OLFM4") %>%
  #subset(grepl("Promoter", annotation)) %>%
  dplyr::select(c(chr, start, end, IGF136819:IGF138590)) %>%
  pivot_longer(-c(chr,start,end), names_to = "sample", values_to = "counts") %>% 
  merge(., oligo_meta, by = "sample") %>% 
  mutate(name = paste0("chr", chr, ":", start, "-", end)) %>%
  ggplot(aes(x = group, y = counts, fill = group)) +
  geom_boxplot(alpha = 0.8) +
  geom_point() +
  facet_wrap(~name, scales = "free_y", nrow = 2) +
  theme_bw() +
  theme(axis.text = element_text(size = 8),
        axis.title.y = element_text(size = 10),
        axis.title.x = element_blank(),
        legend.position = "none", 
        strip.text = element_text(size = 6)) +
  scale_fill_manual(values = c("#00326e", "#e1235c")) +
  scale_y_continuous(name = "Counts") + 
  scale_x_discrete(labels = c("Control", "PD"))


ggsave(oligo_olfm4, filename = here("figures", "20251014_oligo_olfm4.png"),
       height = 10, width = 25, units = "cm", dpi = 600)

```

```{r oligo_go_terms_hyper}
# convert peaks to granges obj
oligo_dds_res_df_hyper_gr <- GRanges(seqnames = oligo_dds_res_df_hyper$chr,
                                     ranges   = IRanges(as.numeric(oligo_dds_res_df_hyper$start), 
                                                        as.numeric(oligo_dds_res_df_hyper$end)),
                                     strand   = "*")

# harmonize chromosome naming style with TxDb (e.g., "chr1" vs "1")
seqlevelsStyle(oligo_dds_res_df_hyper_gr) <- seqlevelsStyle(TxDb.Hsapiens.UCSC.hg38.knownGene)

# annotate peaks
oligo_dds_res_df_hyper_anno <- annotatePeak(oligo_dds_res_df_hyper_gr, 
                                            TxDb=TxDb.Hsapiens.UCSC.hg38.knownGene, 
                                            tssRegion=c(-3000,3000), 
                                            annoDb="org.Hs.eg.db")

# merge annotate to res df
oligo_dds_res_df_hyper <- merge(oligo_dds_res_df_hyper,
                                as.data.frame(oligo_dds_res_df_hyper_anno@anno) %>%
                                  mutate(chr = str_sub(seqnames, 4, 6)),
                                by = c("chr", "start", "end"))

oligo_hyper_genes <- unique(na.omit(as.data.frame(oligo_dds_res_df_hyper_anno)$ENSEMBL))

oligo_hyper_go <- enrichGO(gene=oligo_hyper_genes, 
                           OrgDb=org.Hs.eg.db, 
                           keyType="ENSEMBL",
                           ont="BP", 
                           #pvalueCutoff=0.01, 
                           qvalueCutoff=0.05)
dotplot(oligo_hyper_go)
```

```{r oligo_go_terms_hypo}
# convert peaks to granges obj
oligo_dds_res_df_hypo_gr <- GRanges(seqnames = oligo_dds_res_df_hypo$chr,
                                    ranges   = IRanges(as.numeric(oligo_dds_res_df_hypo$start), 
                                                       as.numeric(oligo_dds_res_df_hypo$end)),
                                    strand   = "*")

# harmonize chromosome naming style with TxDb (e.g., "chr1" vs "1")
seqlevelsStyle(oligo_dds_res_df_hypo_gr) <- seqlevelsStyle(TxDb.Hsapiens.UCSC.hg38.knownGene)

# annotate peaks
oligo_dds_res_df_hypo_anno <- annotatePeak(oligo_dds_res_df_hypo_gr, 
                                           TxDb=TxDb.Hsapiens.UCSC.hg38.knownGene, 
                                           tssRegion=c(-3000,3000), 
                                           annoDb="org.Hs.eg.db")

# merge annotate to res df
oligo_dds_res_df_hypo <- merge(oligo_dds_res_df_hypo,
                               as.data.frame(oligo_dds_res_df_hypo_anno@anno) %>%
                                 mutate(chr = str_sub(seqnames, 4, 6)),
                               by = c("chr", "start", "end"))

oligo_hypo_genes <- unique(na.omit(as.data.frame(oligo_dds_res_df_hypo_anno)$ENSEMBL))

oligo_hypo_go <- enrichGO(gene=oligo_hypo_genes, 
                          OrgDb=org.Hs.eg.db, 
                          keyType="ENSEMBL",
                          ont="BP", 
                          #pvalueCutoff=0.01, 
                          qvalueCutoff=0.05)
dotplot(oligo_hypo_go)
```


### Differential analysis of neurons

```{r neuro_diff_filter}
# remove sex chromosomes
neuro_feat_clean_noxy <- neuro_feat_clean_anno %>%
  subset(!chr %in% c("X", "Y"))

# separate coordinates and counts 
neuro_peak_coords <- neuro_feat_clean_noxy[, c(1:3, 13:27)]
neuro_counts <- as.matrix(neuro_feat_clean_noxy[, -c(1:3, 13:27)])

# recode group as factor
# set control as base level
neuro_meta <- neuro_meta %>%
  mutate(group = relevel(factor(group), 
                         ref = "control")) 

# create edgeR dgelist for filtering
neuro_dgelist <- DGEList(counts = neuro_counts, 
                         samples = neuro_meta)

# replicate-aware filtering - min count 5 
neuro_keep <- filterByExpr(neuro_dgelist, 
                           group = neuro_dgelist$samples$group, 
                           min.count = 5, 
                           min.total.count=50) # smallest group (control) has 4 samples

sum(neuro_keep) # 92886

# remove those with <5 from count and peaks
neuro_peak_coords_clean <- neuro_peak_coords[neuro_keep, ]
neuro_counts_clean <- neuro_counts[neuro_keep, ]

# make deseq2 object
neuro_dds = DESeqDataSetFromMatrix(countData = neuro_counts_clean,
                                   colData = neuro_meta %>%
                                     mutate(pmi_c = scale(pmi, center = T, scale = F)),
                                   design = ~ group + pmi + sex)
```

```{r neuro_check_vif}
md <- as.data.frame(colData(neuro_dds))

# Ensure factors are coded as factors; pmi numeric
md$pd  <- as.factor(md$group)
md$sex <- as.factor(md$sex)
md$pmi <- as.numeric(md$pmi)

# Fit a dummy linear model (response is random noise)
set.seed(1)
fit <- lm(rnorm(nrow(md)) ~ group + pmi + sex, data = md)

# VIFs (GVIF for multi-df factors)
vif(fit)
#  group      pmi      sex 
# 1.036150 1.607185 1.597796 

```

```{r neuro_diff_deseq2}
# diff analysis using deseq2
neuro_dds_diff = DESeq(neuro_dds, 
                       test = "Wald", 
                       fitType = "local",
                       # reduced = ~ sv1, # lrt only
                       sfType = "poscounts")

# get normalized peak counts (uses method from DESeq)
neuro_dds_norm_counts = counts(neuro_dds_diff, 
                               normalized = TRUE)  

# create results dataframe
neuro_dds_res = results(neuro_dds_diff, 
                        pAdjustMethod = "BH",
                        independentFiltering = FALSE, 
                        altHypothesis = "greaterAbs")

# get result names
# resultsNames(neuro_dds_diff)

# # fold change shrinkage
# neuro_dds_res_shr <- lfcShrink(neuro_dds_diff, 
#                                coef = "group_pd_vs_control", 
#                                type = "apeglm")

# make results df
neuro_dds_res_df <- data.frame(
  baseMean = neuro_dds_res$baseMean,
  lfc = neuro_dds_res$log2FoldChange,
  lfcSE = neuro_dds_res$lfcSE,
  # lfc_shr = neuro_dds_res_shr$log2FoldChange,
  stat = neuro_dds_res$stat,
  pvalue = neuro_dds_res$pvalue,
  padj = neuro_dds_res$padj,
  neuro_peak_coords_clean, 
  neuro_dds_norm_counts) 

bc <- bacon(neuro_dds_res_df$stat)
neuro_dds_res_df$padj_bacon <- pval(bc) %>% p.adjust(method = "BH")
neuro_dds_res_df$pvalue_bacon <- pval(bc)
```


Remove age associated regions from oligo pd data
```{r neuro_rm_age}
# convert results ot gr object
neuro_dds_gr <-  makeGRangesFromDataFrame(neuro_dds_res_df %>%
                                            dplyr::select(-any_of(c("seqnames", 
                                                                    "ranges",
                                                                    "strand", 
                                                                    "seqlevels", 
                                                                    "seqlengths", 
                                                                    "isCircular", 
                                                                    "width", 
                                                                    "element"))),
                                          keep.extra.columns = TRUE,
                                          seqnames.field = "chr",
                                          start.field = "start",
                                          end.field = "end")

# put both in same style
seqlevelsStyle(neuro_dds_gr) <- "UCSC"
seqlevelsStyle(neuro_age_sig_gr) <- "UCSC"

# find overlaps with age tt
neuro_overlaps <- findOverlaps(neuro_dds_gr, 
                               neuro_age_sig_gr, 
                               ignore.strand = TRUE) 


# remove overlaps
neuro_dds_res_df_noage <- neuro_dds_res_df[!rownames(neuro_dds_res_df) %in% (neuro_overlaps@from), ]

write.csv(neuro_dds_res_df_noage,
          file = here("data_out",
                      "080_diff_analysis",
                      "20251112_neuro_deseq2_res.csv"))
```

```{r subset_neuro_hyper_hypo}
# get only sig hyper acetylated peaks
neuro_dds_res_df_hyper <- neuro_dds_res_df_noage %>%
  subset(padj < 0.05 & lfc > 0) 

nrow(neuro_dds_res_df_hyper)

# get only sig hypoacetylated peaks
neuro_dds_res_df_hypo <- neuro_dds_res_df_noage %>%
  subset(padj < 0.05 & lfc < 0) 

nrow(neuro_dds_res_df_hypo)
```


```{r neuro_qqplot}
neuro_qqplot <- qq_p_gg(neuro_dds_res_df$pvalue_bacon) +
  # ggtitle("neuroglia") +
  annotate("text", 
           label = paste0("λ=", 
                          round(lambda_from_p(neuro_dds_res_df$pvalue_bacon),2)),
           x = 4, 
           y = 1,
           size = 3)  +
  theme(axis.text = element_text(size = 8),
        axis.title = element_text(size = 10)) 

ggsave(neuro_qqplot, 
       filename = here("figures",
                       "20251112_neuro_qqplot.png"),
       height = 6.5, width = 8, units = "cm", dpi = 600)
```


```{r neuro_big_wigs}
# import oligoglia control and pd bigwigs
neuro_pd_gr <- rtracklayer::import(here("data_out",
                                        "070_peaks", 
                                        "079_bigwigs_groups",
                                        "neuron",
                                        "neuron_disease_merged_cpm.bw"), 
                                   format = "bigWig")

neuro_ctl_gr <- rtracklayer::import(here("data_out",
                                         "070_peaks", 
                                         "079_bigwigs_groups",
                                         "neuron",
                                         "neuron_control_merged_cpm.bw"), 
                                    format = "bigWig")
```

```{r neuro_vis_TSHR_track}
# region (use NCBI-style chr names for hg38)
gen   <- "hg38"
chrom <- 14
start <- 80944815
end   <- 80966364

ax <- GenomeAxisTrack()

mart <- useEnsembl(biomart = "genes", 
                   dataset = "hsapiens_gene_ensembl")

# get gene information for this region
grt <- BiomartGeneRegionTrack(chromosome= chrom,
                              start = start,
                              end = end,
                              genome = gen,
                              name = "Ensembl (HGNC symbols)",
                              biomart = mart)

rg <- ranges(grt)
sym <- mcols(rg)$symbol
gid <- mcols(rg)$gene
sym[is.na(sym) | sym == ""] <- gid[is.na(sym) | sym == ""]
mcols(rg)$symbol <- sym
grt <- GeneRegionTrack(rg, genome = gen, chromosome = chrom, name = "Genes")

# set gene track
displayPars(grt) <- list(
  transcriptAnnotation = "symbol",  # show HGNC symbols
  geneSymbol = TRUE,
  collapseTranscripts = "meta",     # "meta" | "longest" | "gene" | FALSE
  fontsize = 10
)


# get control signal track
control_tr <- DataTrack(
  range       = neuro_ctl_gr,
  genome      = gen,
  chromosome  = chrom,
  type        = "histogram",
  col.histogram="#00326e",
  fill.histogram="#00326e",
  name        = "Control (CPM)",
)

# get pd signal track
pd_tr <- DataTrack(
  range       = neuro_pd_gr,
  genome      = gen,
  chromosome  = chrom,
  type        = "histogram",
  col.histogram="#e1235c", 
  fill.histogram="#e1235c",
  name        = "PD (CPM)"
)

# create track to indicate where peak
peak_start <- 80954815
peak_end   <- 80956364

# Build a clean GRanges with an 'id' column (character, length == n features)
diff_peak <- GRanges(seqnames = chrom, ranges = IRanges(peak_start, peak_end))
mcols(diff_peak)$id <- NA  # label to display

# Sanity check: id length matches feature count
stopifnot(length(mcols(diff_peak)$id) == length(diff_peak))

diff_tr <- AnnotationTrack(
  range      = diff_peak,
  genome     = gen,
  chromosome = chrom,
  col        = "black",
  fill = "black",
  lwd        = 2,
  stacking   = "dense"
)

displayPars(diff_tr) <- c(
  displayPars(diff_tr),
  list(
    featureAnnotation = NULL,  # draw text from mcols(...)$id
    fontsize          = 9,
    just              = "above"  # place label above feature
  )
)

# get chromosome track
ideo_tr <- IdeogramTrack(genome = gen, chromosome = chrom)

# plot tracks together 
png(here("figures", "neuro_tshr_track.png"), width = 3600, height = 3600, res = 600)

plotTracks(list(ideo_tr, 
                ax, 
                grt, 
                pd_tr, 
                control_tr, 
                diff_tr),
           chromosome = chrom, 
           from = start,
           to = end,
           background.title = "white",
           col.title = "black",
           cex.axis = 0.6,   # bigger axis tick labels
           cex.title = 0.8,  # track titles
           cex.main = 0.8,
            sizes = c(0.25, 1, 1, 2, 2, 0.25))    # if you use main = "...")

dev.off()
```

```{r neuro_vis_ADARB2_boxlot}
neuro_dds_res_df_noage %>%
  subset(padj_bacon < 0.05) %>%
  arrange(padj_bacon) %>%
  slice_head(n=1) %>%
  dplyr::select("IGF136818":"IGF138589") %>%
  pivot_longer(everything(),
               names_to = "sample",
               values_to = "counts") %>%
  merge(., 
        neuro_meta,
        by = "sample") %>%
  ggplot(aes(x=group,
             y = counts, 
             fill = group)) +
  geom_boxplot(alpha = 0.8) +
  geom_point() +
  theme_bw() +
  theme(axis.text = element_text(size = 8),
        axis.title.y = element_text(size = 10),
        axis.title.x = element_blank(),
        legend.position = "none") +
  scale_fill_manual(values = c("#00326e", "#e1235c"))
```

```{r neuro_volcano}
neuro_volcano <- ggplot(neuro_dds_res_df, 
                        aes(x = lfc, 
                            y = -log(pvalue_bacon),
                            color = ifelse(padj_bacon < 0.05,"FDR<0.05", "FDR≥0.05"))) +
  geom_point(alpha = 0.8,
             size = 1) +
  theme_bw() +
  scale_color_manual(values = c("#e1235c", "#00326e"),
                     name = "Significance") +
  scale_y_continuous(name = "-log(unadjusted p-value)") +
  scale_x_continuous(name = "log(fold-change)") +
  theme(axis.text = element_text(size = 8),
        axis.title = element_text(size = 10),
        legend.title = element_text(size = 9),
        legend.text = element_text(size = 8), 
        legend.key.height = unit(0.25, "cm"),
        legend.key.width = unit(0.25, "cm"),
        legend.position = "bottom") +
  ggrepel::geom_text_repel(data = neuro_dds_res_df %>% 
                             arrange(padj_bacon) %>% 
                             slice_head(n=10),
                           aes(label = SYMBOL),
                           max.overlaps = 100,
                           box.padding = 0.4, 
                           min.segment.length = unit(0, 'lines'),
                           color = "black", 
                           size = 2.5) 

ggsave(neuro_volcano, 
       filename = here("figures",
                       "20251112_neuro_volcano_legend.png"),
       height = 8, width = 8, units = "cm", dpi = 600)
```


```{r neuro_recap_rat_rotenone}
# bcl 
neuro_bcl <- neuro_dds_res_df %>% 
  subset(SYMBOL == "BCL3") %>%
  subset(grepl("Promoter", annotation)) %>%
  dplyr::select(c(chr, start, end, IGF136819:IGF138590)) %>%
  pivot_longer(-c(chr,start,end), names_to = "sample", values_to = "counts") %>% 
  merge(., neuro_meta, by = "sample") %>% 
  mutate(name = paste0("chr", chr, ":", start, "-", end)) %>%
  ggplot(aes(x = group, y = counts, fill = group)) +
  geom_boxplot(alpha = 0.8) +
  geom_point() +
  facet_wrap(~name, scales = "free_y", nrow = 1) +
  theme_bw() +
  theme(axis.text = element_text(size = 8),
        axis.title.y = element_text(size = 10),
        axis.title.x = element_blank(),
        legend.position = "none", 
        strip.text = element_text(size = 6)) +
  scale_fill_manual(values = c("#00326e", "#e1235c")) +
  scale_y_continuous(name = "Counts") + 
  scale_x_discrete(labels = c("Control", "PD"))

ggsave(neuro_bcl, filename = here("figures", "20251014_neuro_bcl3.png"),
       height = 5, width = 15, units = "cm", dpi = 600)

# mat2b 
neuro_mat2b <- neuro_dds_res_df %>% 
  subset(SYMBOL == "MAT2B") %>%
  subset(grepl("Distal Intergenic", annotation)) %>%
  dplyr::select(c(chr, start, end, IGF136819:IGF138590)) %>%
  pivot_longer(-c(chr,start,end), names_to = "sample", values_to = "counts") %>% 
  merge(., neuro_meta, by = "sample") %>% 
  mutate(name = paste0("chr", chr, ":", start, "-", end)) %>%
  ggplot(aes(x = group, y = counts, fill = group)) +
  geom_boxplot(alpha = 0.8) +
  geom_point() +
  facet_wrap(~name, scales = "free_y") +
  theme_bw() +
  theme(axis.text = element_text(size = 8),
        axis.title.y = element_text(size = 10),
        axis.title.x = element_blank(),
        legend.position = "none", 
        strip.text = element_text(size = 6)) +
  scale_fill_manual(values = c("#00326e", "#e1235c")) +
  scale_y_continuous(name = "Counts") + 
  scale_x_discrete(labels = c("Control", "PD"))

ggsave(neuro_mat2b, filename = here("figures", "20251014_neuro_mat2b.png"),
       height = 5, width = 5, units = "cm", dpi = 600)

# hif3a 
neuro_hif3a <- neuro_dds_res_df %>% 
  subset(SYMBOL == "HIF3A") %>%
  subset(grepl("Promoter", annotation)) %>%
  dplyr::select(c(chr, start, end, IGF136819:IGF138590)) %>%
  pivot_longer(-c(chr,start,end), names_to = "sample", values_to = "counts") %>% 
  merge(., neuro_meta, by = "sample") %>% 
  mutate(name = paste0("chr", chr, ":", start, "-", end)) %>%
  ggplot(aes(x = group, y = counts, fill = group)) +
  geom_boxplot(alpha = 0.8) +
  geom_point() +
  facet_wrap(~name, scales = "free_y") +
  theme_bw() +
  theme(axis.text = element_text(size = 8),
        axis.title.y = element_text(size = 10),
        axis.title.x = element_blank(),
        legend.position = "none", 
        strip.text = element_text(size = 6)) +
  scale_fill_manual(values = c("#00326e", "#e1235c")) +
  scale_y_continuous(name = "Counts") + 
  scale_x_discrete(labels = c("Control", "PD"))

ggsave(neuro_hif3a, filename = here("figures", "20251014_neuro_hif3a.png"),
       height = 5, width = 15, units = "cm", dpi = 600)

# olfm4 
neuro_olfm4 <- neuro_dds_res_df %>% 
  subset(SYMBOL == "OLFM4") %>%
  #subset(grepl("Promoter", annotation)) %>%
  dplyr::select(c(chr, start, end, IGF136819:IGF138590)) %>%
  pivot_longer(-c(chr,start,end), names_to = "sample", values_to = "counts") %>% 
  merge(., neuro_meta, by = "sample") %>% 
  mutate(name = paste0("chr", chr, ":", start, "-", end)) %>%
  ggplot(aes(x = group, y = counts, fill = group)) +
  geom_boxplot(alpha = 0.8) +
  geom_point() +
  facet_wrap(~name, scales = "free_y", nrow = 2) +
  theme_bw() +
  theme(axis.text = element_text(size = 8),
        axis.title.y = element_text(size = 10),
        axis.title.x = element_blank(),
        legend.position = "none", 
        strip.text = element_text(size = 6)) +
  scale_fill_manual(values = c("#00326e", "#e1235c")) +
  scale_y_continuous(name = "Counts") + 
  scale_x_discrete(labels = c("Control", "PD"))


ggsave(neuro_olfm4, filename = here("figures", "20251014_neuro_olfm4.png"),
       height = 10, width = 25, units = "cm", dpi = 600)

```

```{r neuro_go_terms_hyper}
# convert peaks to granges obj
neuro_dds_res_df_hyper_gr <- GRanges(seqnames = neuro_dds_res_df_hyper$chr,
                                     ranges   = IRanges(as.numeric(neuro_dds_res_df_hyper$start), 
                                                        as.numeric(neuro_dds_res_df_hyper$end)),
                                     strand   = "*")

# harmonize chromosome naming style with TxDb (e.g., "chr1" vs "1")
seqlevelsStyle(neuro_dds_res_df_hyper_gr) <- seqlevelsStyle(TxDb.Hsapiens.UCSC.hg38.knownGene)

# annotate peaks
neuro_dds_res_df_hyper_anno <- annotatePeak(neuro_dds_res_df_hyper_gr, 
                                            TxDb=TxDb.Hsapiens.UCSC.hg38.knownGene, 
                                            tssRegion=c(-3000,3000), 
                                            annoDb="org.Hs.eg.db")

# merge annotate to res df
neuro_dds_res_df_hyper <- merge(neuro_dds_res_df_hyper,
                                as.data.frame(neuro_dds_res_df_hyper_anno@anno) %>%
                                  mutate(chr = str_sub(seqnames, 4, 6)),
                                by = c("chr", "start", "end"))

neuro_hyper_genes <- unique(na.omit(as.data.frame(neuro_dds_res_df_hyper_anno)$ENSEMBL))

neuro_hyper_go <- enrichGO(gene=neuro_hyper_genes, 
                           OrgDb=org.Hs.eg.db, 
                           keyType="ENSEMBL",
                           ont="BP", 
                           #pvalueCutoff=0.01, 
                           qvalueCutoff=0.05)
dotplot(neuro_hyper_go)
```

```{r neuro_go_terms_hypo}
# convert peaks to granges obj
neuro_dds_res_df_hypo_gr <- GRanges(seqnames = neuro_dds_res_df_hypo$chr,
                                    ranges   = IRanges(as.numeric(neuro_dds_res_df_hypo$start), 
                                                       as.numeric(neuro_dds_res_df_hypo$end)),
                                    strand   = "*")

# harmonize chromosome naming style with TxDb (e.g., "chr1" vs "1")
seqlevelsStyle(neuro_dds_res_df_hypo_gr) <- seqlevelsStyle(TxDb.Hsapiens.UCSC.hg38.knownGene)

# annotate peaks
neuro_dds_res_df_hypo_anno <- annotatePeak(neuro_dds_res_df_hypo_gr, 
                                           TxDb=TxDb.Hsapiens.UCSC.hg38.knownGene, 
                                           tssRegion=c(-3000,3000), 
                                           annoDb="org.Hs.eg.db")

# merge annotate to res df
neuro_dds_res_df_hypo <- merge(neuro_dds_res_df_hypo,
                               as.data.frame(neuro_dds_res_df_hypo_anno@anno) %>%
                                 mutate(chr = str_sub(seqnames, 4, 6)),
                               by = c("chr", "start", "end"))

neuro_hypo_genes <- unique(na.omit(as.data.frame(neuro_dds_res_df_hypo_anno)$ENSEMBL))

neuro_hypo_go <- enrichGO(gene=neuro_hypo_genes, 
                          OrgDb=org.Hs.eg.db, 
                          keyType="ENSEMBL",
                          ont="BP", 
                          #pvalueCutoff=0.01, 
                          qvalueCutoff=0.05)
dotplot(neuro_hypo_go)
```


### Comparison of peaks across cell types

```{r upset}
# Make disjoint tiles that partition the union of all ranges
clusters <- reduce(c(micro_dds_gr, oligo_dds_gr, neuro_dds_gr), ignore.strand = TRUE)

# Membership matrix: does a tile overlap each set?
mats <- list(
  countOverlaps(clusters, micro_dds_gr, ignore.strand = F) > 0,
  countOverlaps(clusters, oligo_dds_gr, ignore.strand = F) > 0,
  countOverlaps(clusters, neuro_dds_gr, ignore.strand = F) > 0
)

memb <- as.data.frame(setNames(mats, c("microglia", "oligodendrocytes", "neurons")))


upset <-  upset(
  memb,
  intersect = c("microglia", "oligodendrocytes", "neurons"),
  stripes=upset_stripes(
    geom=geom_segment(size=7),
    colors=c("#00326e", "#e1235c", "#63dcff")
  ),
  base_annotations = list(
    "Intersection size" = intersection_size(
      counts = TRUE,
      text = list(size = 2),
      theme = theme(axis.text.y = element_text(size = 7)))),
  matrix = intersection_matrix(
    geom = geom_point(size = 2)
  ),
  set_sizes=FALSE,
  height_ratio = 0.30 
) 


ggsave(upset, filename = here("figures",
                              "20251014_celltype_counts_comparison.png"),
       height = 8, width = 8, units = "cm", dpi = 600)
```


### Output size factors

```{r size factors}
# get size factors
micro_size_factors <- sizeFactors(micro_dds_diff) %>%
  as.data.frame() %>%
  rownames_to_column() %>%
  merge(.,
        micro_meta %>% 
          dplyr::select(sample, celltype, group),
        by.x = "rowname", by.y= "sample")

oligo_size_factors <- sizeFactors(oligo_dds_diff) %>%
  as.data.frame() %>%
  rownames_to_column() %>%
  merge(.,
        oligo_meta %>% 
          dplyr::select(sample, celltype, group),
        by.x = "rowname", by.y= "sample")

neuro_size_factors <- sizeFactors(neuro_dds_diff) %>%
  as.data.frame() %>%
  rownames_to_column() %>%
  merge(.,
        neuro_meta %>% 
          dplyr::select(sample, celltype, group),
        by.x = "rowname", by.y= "sample")

size_factors <- rbind(micro_size_factors,
                      oligo_size_factors,
                      neuro_size_factors) %>%
  as.data.frame() 

write_tsv(size_factors,
          here("data_out",
               "080_differential_analysis",
               "size_factors.tsv"),
          col_names = FALSE)
```


